{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Contains all the code for generating figures in the publication. Requires filtered NanoSeq mutations, callable coverage arrays, and MA line mutations and 1001 variants (from reformatting_ma_line_muts.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "sys.path.append(os.getcwd() + '/../python_scripts') # this lets us import files in python_scripts (like gtools)\n",
    "import gtools\n",
    "if os.getcwd()[:8] != '/scratch': # switch to the directory where all the data files are\n",
    "    os.chdir(f'/scratch/cam02551/{os.getcwd().split(\"/\")[-2]}')\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import csv\n",
    "from collections import defaultdict\n",
    "from tqdm import tqdm\n",
    "import numpy as np\n",
    "import pysam\n",
    "import subprocess\n",
    "import importlib\n",
    "import random\n",
    "%matplotlib inline\n",
    "import matplotlib.patches as patches\n",
    "import pickle\n",
    "import math\n",
    "import scipy.stats as stats\n",
    "import scipy.cluster as cluster\n",
    "plt.rcParams['svg.fonttype'] = 'none'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_names = ''\n",
    "sample_names += 'untreated_1 untreated_2 untreated_3 untreated_4 untreated_5 untreated_6 untreated_7 untreated_8 '\n",
    "# sample_names += 'UVB_1 UVB_2 UVC_002 UVC_005 UVC_01 UVC_02 '\n",
    "# sample_names += 'Ler-0_1 '\n",
    "sample_names = sample_names.split()\n",
    "\n",
    "mut_files = [f'data/variant/{s}_merged_filtered.tsv' for s in sample_names] # filtered mutations from snakemake pipeline\n",
    "cov_prefixes = {s:f'data/coverage/{s}_merged_duplex/' for s in sample_names} # callable coverage array prefixes from snakemake pipeline\n",
    "genome_fasta = 'data/ref/ref.fa'\n",
    "# conflict rates can be found in the stdout of duplex_metadata.py\n",
    "conflict_rate_dict = {'untreated_1_merged.bam': 0.1421243306017736, 'untreated_2_merged.bam': 0.07499291504028818, 'untreated_3_merged.bam': 0.06698923466207254, 'untreated_4_merged.bam': 0.08825301441514141, 'untreated_5_merged.bam': 0.10694477979171899, 'untreated_6_merged.bam': 0.06708132263901942, 'untreated_7_merged.bam': 0.09039258234998046, 'untreated_8_merged.bam': 0.1131217594141343, 'UVB_1_merged.bam': 0.005914724117601422, 'UVB_2_merged.bam': 0.02487706726801387, 'UVC_002_merged.bam': 0.007101074947091077, 'UVC_005_merged.bam': 0.012911711246730054, 'UVC_01_merged.bam': 0.005330398460813733, 'UVC_02_merged.bam': 0.014511010508950055, 'Ler-0_1_merged.bam': 0.03273057807000435}\n",
    "ma_line_dir = 'data/variant/weng_split/' # directory containing a tsv variant file for each MA line (generated in reformatting_ma_line_muts.ipynb)\n",
    "polymorph_file = 'data/variant/1001_muts.tsv' # 1001 genomes polymorphisms (generated in reformatting_ma_line_muts.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "conflict_rates = [conflict_rate_dict[s + '_merged.bam'] for s in sample_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs('figs/', exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotypes = list({s.rsplit('_', 1)[0]:0 for s in sample_names})\n",
    "\n",
    "# convert a sample name to its genotype name\n",
    "def s_to_g(sample):\n",
    "    return sample.rsplit('_', 1)[0]\n",
    "\n",
    "def g_to_s(genotype):\n",
    "    return [s for s in sample_names if s_to_g(s) == genotype]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gen_cmap(gen):\n",
    "    gen_to_i = {'untreated':0, 'UVB':2, 'UVC':3, 'MA lines':6, 'Col-0-swapped':11}\n",
    "    if gen in gen_to_i:\n",
    "        return gtools.base_cmap[gen_to_i[gen]]\n",
    "    \n",
    "    gen_to_i = {'Ler-0':8, '1001':9}\n",
    "    if gen in gen_to_i:\n",
    "        return gtools.light_cmap[gen_to_i[gen]]\n",
    "\n",
    "    return 'grey'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load genome fasta\n",
    "genome = gtools.load_genome(genome_fasta)\n",
    "chrom_sizes = {chrom:len(genome[chrom]) for chrom in genome if chrom != 'ChrC' and chrom != 'ChrM'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load the mutations\n",
    "remove duplicates, non-nuclear muts, and contaminants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the filtered mutations, discard duplicates and non-nuclear contigs\n",
    "dfs_mut = dict()\n",
    "for i, file in enumerate(mut_files):\n",
    "    df = pd.read_table(file, quoting=csv.QUOTE_NONE)\n",
    "    raw_num = len(df)\n",
    "    \n",
    "    df = df.drop_duplicates('chrom pos ref alt'.split())\n",
    "    unique_num = len(df)\n",
    "    \n",
    "    df = df[(df.chrom != 'ChrC') & (df.chrom != 'ChrM')]\n",
    "    df['sample_freq'] = df.sample_sup / df.sample_cov\n",
    "    df.sample_freq = df.sample_freq.fillna(0)\n",
    "    print(f'{sample_names[i]}: {raw_num} muts loaded, {unique_num} unique, {len(df)} nuclear' + (f', WARNING {sum(df.sample_cov == 0)} with undefined frequency set to 0' if sum(df.sample_cov == 0) > 0 else ''))\n",
    "    \n",
    "    dfs_mut[sample_names[i]] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# label variants which are present in a sibling (same genotype)\n",
    "for g in genotypes:\n",
    "    for s1 in g_to_s(g):\n",
    "        in_sibling = np.zeros(len(dfs_mut[s1]), dtype=bool)\n",
    "        for s2 in g_to_s(g):\n",
    "            if s1 == s2:\n",
    "                continue\n",
    "            df = dfs_mut[s1].merge(dfs_mut[s2], how='left', on='chrom pos ref alt'.split(), indicator='_merge')\n",
    "            in_sibling = in_sibling + np.array(df._merge == 'both')\n",
    "        dfs_mut[s1]['in_sibling'] = in_sibling\n",
    "        print(f'{s1}: {sum(in_sibling)} muts in a sibling')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each pair of samples, get all the \"contaminant\" mutations found in both\n",
    "dfs_contam = []\n",
    "for sample_1 in sample_names:\n",
    "    for sample_2 in sample_names:\n",
    "        if s_to_g(sample_1) == s_to_g(sample_2): # ignore siblings\n",
    "            continue\n",
    "        if len(dfs_mut[sample_1]) > 10000 or len(dfs_mut[sample_2]) > 10000: # ignore samples with large numbers of mutations (the Ler-0 germline mutations)\n",
    "            continue\n",
    "        if 'swapped' in sample_1 or 'swapped' in sample_2: # ignore swapped samples\n",
    "            continue\n",
    "        \n",
    "        df = dfs_mut[sample_1].merge(dfs_mut[sample_2]['chrom pos alt ref'.split()], how='inner')\n",
    "        if len(df) > 0:\n",
    "            # print(f'{sum(df.in_sibling)} germline and {sum(~df.in_sibling)} somatic mutations in {sample_1} are in {sample_2}')\n",
    "            dfs_contam.append(df.copy())\n",
    "df_contam = pd.concat(dfs_contam) if len(dfs_contam) > 0 else pd.DataFrame(columns='chrom pos ref alt'.split())\n",
    "df_contam = df_contam.sort_values('chrom pos ref alt'.split())\n",
    "contaminants = set(zip(df_contam.chrom, df_contam.pos, df_contam.ref, df_contam.alt))\n",
    "print(f'{len(contaminants)} unique and {len(df_contam)} total contaminant mutations')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove all contaminant mutations\n",
    "for s in sample_names:\n",
    "    if 'swapped' in s or len(dfs_mut[s]) > 10000: # don't discard \"contaminants\" for the swapped samples or samples with a large number of muts\n",
    "        continue\n",
    "\n",
    "    before_num = len(dfs_mut[s])\n",
    "    dfs_mut[s] = dfs_mut[s][dfs_mut[s].apply(lambda r: (r.chrom, r.pos, r.ref, r.alt) not in contaminants, axis=1)]\n",
    "    print(f'{s}: removed {before_num - len(dfs_mut[s])} contaminants')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Separate somatic and germline muts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs_mut_g = {g:None for g in genotypes}\n",
    "for s in dfs_mut:\n",
    "    if dfs_mut_g[s_to_g(s)] is None:\n",
    "        dfs_mut_g[s_to_g(s)] = dfs_mut[s].copy()\n",
    "    else:\n",
    "        dfs_mut_g[s_to_g(s)] = pd.concat([dfs_mut_g[s_to_g(s)], dfs_mut[s]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vis_genotypes = genotypes\n",
    "\n",
    "# plot histograms of mutation freq, color based on presence in sibling or another fragment\n",
    "# bins = np.append(np.linspace(0, 0.014, 15), np.linspace(0.015, 1.015, 25))\n",
    "bins = np.logspace(np.log10(0.001), np.log10(1), 30)\n",
    "breakpoint = 15\n",
    "fig, axs = plt.subplots(math.ceil(len(vis_genotypes) / 3), min(len(vis_genotypes), 3), sharex=True, figsize=(4 * min(len(vis_genotypes), 3), 0.5 * math.ceil((len(vis_genotypes) + 3) / 3)))\n",
    "for i, g in enumerate(vis_genotypes):\n",
    "    df = dfs_mut_g[g]\n",
    "    if len(vis_genotypes) == 1:\n",
    "        ax = axs\n",
    "    elif len(vis_genotypes) <= 3:\n",
    "        ax = axs[i % 3]\n",
    "    else:\n",
    "        ax = axs[i // 3][i % 3]\n",
    "    cmap = plt.get_cmap('tab20')\n",
    "    \n",
    "    hml = ax.hist(df[~df.in_sibling & (df.sample_sup > 2) & (df.sample_freq < 0.35)].sample_freq, bins=bins, rwidth=0.85, color=cmap(2))\n",
    "    hmh = ax.hist(df[~df.in_sibling & (df.sample_sup > 2) & (df.sample_freq >= 0.35)].sample_freq, bins=bins, rwidth=0.85, color=cmap(3), bottom=hml[0])\n",
    "    hul = ax.hist(df[~df.in_sibling & (df.sample_sup <= 2) & (df.sample_freq < 0.35)].sample_freq, bins=bins, rwidth=0.85, color=cmap(0), bottom=hml[0] + hmh[0])\n",
    "    huh = ax.hist(df[~df.in_sibling & (df.sample_sup <= 2) & (df.sample_freq >= 0.35)].sample_freq, bins=bins, rwidth=0.85, color=cmap(1), bottom=hml[0] + hmh[0] + hul[0])\n",
    "    hs = ax.hist(df[df.in_sibling].sample_freq, bins=bins, rwidth=0.85, color=cmap(5), bottom=hul[0] + huh[0] + hml[0] + hmh[0])\n",
    "    \n",
    "    ax.set_xscale('log')\n",
    "    ax.set_title(g, y=0.45)\n",
    "ax.set_xlabel('Frequency')\n",
    "fig.legend([hul[2], huh[2], hml[2], hmh[2], hs[2]], ['1 fragment', '1 framgnet (filtered as germline)', '>1 fragment', '>1 fragment (filtered as germline)', 'in sibling (filtered as germline)'], bbox_to_anchor=(0.9, 0.91), loc='lower right')\n",
    "fig.suptitle('Histogram of mutation frequencies')\n",
    "fig.savefig('figs/hist_of_mutation_frequencies.png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# identify somatic mutations as those not present in a sibling, and at low frequency\n",
    "dfs_somatic = dict()\n",
    "for sample in dfs_mut:\n",
    "    df = dfs_mut[sample]\n",
    "    if sum(df.sample_freq.isna()) > 0:\n",
    "        print(f'WARNING: {sum(df.sample_freq.isna())} mutations in {sample} with undefined frequency, assigning these a frequency of 0')\n",
    "        df.sample_freq = df.sample_freq.fillna(0)\n",
    "    if 'swapped' in sample: # don't discard swapped sample muts if in a \"sibling\"\n",
    "        dfs_somatic[sample] = df[df.sample_freq < 0.35].copy()\n",
    "    elif 'Ler-0' in sample: # only keep high abundance Ler-0 mutations\n",
    "        dfs_somatic[sample] = df[df.sample_freq > 0.35].copy()\n",
    "    else: # keep only low-abundance unique mutations as somatic\n",
    "        dfs_somatic[sample] = df[~df.in_sibling & (df.sample_freq < 0.35)].copy()\n",
    "    print(f'{sample}: kept {len(dfs_somatic[sample])} of {len(df)} mutations as \"somatic\"')\n",
    "    \n",
    "# # identify germline mutations as those which are present in a sibling\n",
    "# dfs_germline = dict()\n",
    "# for sample in dfs_mut:\n",
    "#     df = dfs_mut[sample]\n",
    "#     dfs_germline[sample] = df[df.in_sibling | (df.sample_freq >= 0.35)].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# add the MA line and 1001 polymorphism data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'MA lines' not in genotypes:\n",
    "    for file in os.listdir(ma_line_dir):\n",
    "        df = pd.read_table(f'{ma_line_dir}{file}')\n",
    "        df = df[df.alt != 'N']\n",
    "        dfs_somatic[f'MA lines_{file.split(\"_\")[1]}'] = df\n",
    "        sample_names.append(f'MA lines_{file.split(\"_\")[1]}')\n",
    "        conflict_rates.append(0)\n",
    "    genotypes += ['MA lines']\n",
    "    print(f'loaded {sum([len(dfs_somatic[sample]) for sample in sample_names if s_to_g(sample) == \"MA lines\"])} MA line mutations')\n",
    "    \n",
    "    dfs_somatic['1001_0'] = pd.read_table(polymorph_file)\n",
    "    sample_names.append('1001_0')\n",
    "    genotypes.append('1001')\n",
    "    conflict_rates.append(0)\n",
    "    print(f'loaded {len(dfs_somatic[\"1001_0\"])} 1001 mutations')\n",
    "else:\n",
    "    print('ERROR: rerunning this cell requires rerunning the header cells first')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Load callable coverage data (always run, it alters the MA line and 1001 datasets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_call_cov = {sample:0 for sample in sample_names}\n",
    "for chrom in chrom_sizes:\n",
    "    for i, sample in tqdm(enumerate(sample_names), total=len(sample_names)):\n",
    "        if 'MA lines' in sample or '1001' in sample:\n",
    "            if i == 0 or 'MA lines' not in sample_names[i - 1]: # if the last sample wasn't also an MA line sample, this if statement just saves time\n",
    "                arr = np.load(f'{cov_prefixes['untreated_1']}{chrom}.npy')\n",
    "                arr = arr > 0\n",
    "            \n",
    "            # remove any mutations in regions I'm defining as uncallable (this does discard about 300/2600 MA line mutations though)\n",
    "            dfs_somatic[sample] = dfs_somatic[sample][(dfs_somatic[sample].chrom != chrom) | dfs_somatic[sample].pos.map(lambda x: arr[x] if x < len(arr) else True)]\n",
    "        elif 'Ler-0' in sample:\n",
    "            arr = np.load(f'{cov_prefixes[sample]}{chrom}.npy')\n",
    "            arr = arr > 0\n",
    "        else:\n",
    "            arr = np.load(f'{cov_prefixes[sample]}{chrom}.npy')\n",
    "        sample_call_cov[sample] += np.sum(arr) * (1 - conflict_rates[i]) # FIXME is it safe to do the conflict rate correction here?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cov_samples = sample_names[:8]\n",
    "callable_histogram = {sample:np.zeros(1000, dtype=int) for sample in cov_samples}\n",
    "for sample in tqdm(cov_samples):\n",
    "    for chrom in chrom_sizes:\n",
    "        arr = np.load(f'{cov_prefixes[sample]}{chrom}.npy')\n",
    "        callable_histogram[sample] += np.histogram(arr, bins=range(1001))[0]\n",
    "print('number of sites with at least one callable coverage:')\n",
    "for sample in cov_samples:\n",
    "    print(f'{sample}: {np.sum(callable_histogram[sample][1:])}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(len(cov_samples), sharex=True)\n",
    "x_max = 150\n",
    "for i, sample in enumerate(cov_samples):\n",
    "    axs[i].bar(range(x_max), callable_histogram[sample][:x_max])\n",
    "    axs[i].set_title(sample, y=0.3, loc='right')\n",
    "axs[-1].set_xlabel('Callable coverage')\n",
    "axs[len(axs) // 2].set_ylabel('Count (genomic sites)')\n",
    "plt.show()\n",
    "fig.savefig('figs/callable_coverage_histogram.png', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save mutations for supplemental table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save mutations\n",
    "df_supp_table = pd.DataFrame(columns='sample chrom pos ref alt'.split())\n",
    "for s in sample_names:\n",
    "    if s_to_g(s) not in ['Col-0', 'UVB', 'UVC']:\n",
    "        continue\n",
    "    df = dfs_somatic[s]['chrom pos ref alt'.split()].copy()\n",
    "    correct_name = f'untreated-{s[-1]}' if s_to_g(s) == 'Col-0' else s\n",
    "    df['sample'] = [correct_name] * len(df)\n",
    "    df_supp_table = pd.concat([df_supp_table, df])\n",
    "\n",
    "df_supp_table.to_csv('data/supp_table.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Investigate mutation frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ignore_samples = ['Ler-0', 'MA lines', '1001']\n",
    "vis_samples = [s for s in sample_names if sum([x in s for x in ignore_samples]) == 0]\n",
    "vis_genotypes = list({s_to_g(s):0 for s in vis_samples})\n",
    "\n",
    "# TODO I need something that conveys what this plot is trying to do while correcting for coverage\n",
    "# count the number of mutants observed at least 1, 2, 3, 4, and 5 times\n",
    "max_obs = 5\n",
    "obs_sample_counts = np.zeros((max_obs, len(vis_samples))) # obs_rates[number of observations][sample index] = fraction of mutations observed at least that many times + 1\n",
    "obs_genotype_counts = np.zeros((max_obs, len(vis_genotypes))) \n",
    "for i in range(max_obs):\n",
    "    for j, sample in enumerate(vis_samples):\n",
    "        obs_sample_counts[i][j] = sum((dfs_somatic[sample].sample_sup / 2) > i)\n",
    "        obs_genotype_counts[i][vis_genotypes.index(s_to_g(sample))] += sum((dfs_somatic[sample].sample_sup / 2) > i)\n",
    "\n",
    "obs_sample_rates = obs_sample_counts / obs_sample_counts[0]\n",
    "obs_genotype_rates = obs_genotype_counts / obs_genotype_counts[0]\n",
    "\n",
    "        \n",
    "fig, ax = plt.subplots(figsize=(12, 3))\n",
    "# h = np.zeros(len(vis_samples))\n",
    "for i in range(max_obs):\n",
    "    ax.bar(range(len(vis_samples)), obs_sample_rates[i]) # don't plot ler\n",
    "    # h += obs_sample_rates[i]\n",
    "ax.legend([f'>={max_obs} observations' if i == max_obs else f'{i} observation(s)' for i in range(1, max_obs + 1)], loc='upper left', bbox_to_anchor=(0, 1))\n",
    "ax.set_xticks(range(len(vis_samples)))\n",
    "# ax.set_ylim(0, 0.3)\n",
    "ax.set_xticklabels(vis_samples, rotation=90)\n",
    "ax.set_title('Fraction of somatic mutations observed in multiple fragments')\n",
    "fig.savefig('figs/sfs_stacked_bar_samples.png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(6, 3))\n",
    "h = np.zeros(len(vis_genotypes))\n",
    "for i in range(max_obs):\n",
    "    ax.bar(range(len(vis_genotypes)), obs_genotype_rates[i], bottom=h)\n",
    "    h += obs_genotype_rates[i]\n",
    "ax.legend([f'>={max_obs} observations' if i == max_obs else f'{i} observation(s)' for i in range(1, max_obs + 1)])\n",
    "ax.set_xticks(range(len(vis_genotypes)), labels=vis_genotypes)\n",
    "ax.set_title('Fraction of somatic mutations observed in multiple fragments')\n",
    "fig.savefig('figs/sfs_stacked_bar_genotypes.png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "fig, ax = plt.subplots()\n",
    "for c, genotype in enumerate(vis_genotypes):\n",
    "    x = []\n",
    "    y = []\n",
    "    for sample in vis_samples:\n",
    "        if s_to_g(sample) == genotype:\n",
    "            x.append(sample_call_cov[sample])\n",
    "            y.append(sum(dfs_somatic[sample].sample_sup > 2) / sum(dfs_somatic[sample].sample_cov > 2))\n",
    "    ax.scatter(x, y, color=gen_cmap(c))\n",
    "ax.legend(vis_genotypes, loc='upper left', bbox_to_anchor=(1, 1))\n",
    "ax.set_xlabel('Callable coverage')\n",
    "ax.set_ylabel('Fraction of somatic mutations\\nobserved in >1 fragment')\n",
    "ax.set_title('Fraction \"high frequency\" somatic mutations')\n",
    "fig.savefig('figs/frac_high_freq_somatic_mutations_call_cov.png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "fig, ax = plt.subplots()\n",
    "for c, genotype in enumerate(vis_genotypes):\n",
    "    x = []\n",
    "    y = []\n",
    "    for sample in vis_samples:\n",
    "        if s_to_g(sample) == genotype:\n",
    "            x.append(len(dfs_somatic[sample]) / sample_call_cov[sample])\n",
    "            y.append(sum(dfs_somatic[sample].sample_sup > 2) / sum(dfs_somatic[sample].sample_cov > 2))\n",
    "    ax.scatter(x, y, color=gen_cmap(c))\n",
    "ax.legend(vis_genotypes, loc='upper left', bbox_to_anchor=(1, 1))\n",
    "ax.set_xlabel('Somatic mutation rate')\n",
    "ax.set_ylabel('Fraction of somatic mutations\\nobserved in >1 fragment')\n",
    "ax.set_title('Fraction \"high frequency\" somatic mutations')\n",
    "fig.savefig('figs/frac_high_freq_somatic_mutations_mut_rate.png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "fig, ax = plt.subplots()\n",
    "for c, genotype in enumerate(vis_genotypes):\n",
    "    x = []\n",
    "    y = []\n",
    "    for sample in sample_names:\n",
    "        if s_to_g(sample) == genotype:\n",
    "            x.append(len(dfs_somatic[sample]))\n",
    "            y.append(sum(dfs_somatic[sample].sample_sup > 2) / sum(dfs_somatic[sample].sample_cov > 2))\n",
    "    ax.scatter(x, y, color=gen_cmap(c))\n",
    "ax.legend(vis_genotypes, loc='upper left', bbox_to_anchor=(1, 1))\n",
    "ax.set_xlabel('Somatic mutations detected')\n",
    "ax.set_ylabel('Fraction of somatic mutations\\nobserved in >1 fragment')\n",
    "ax.set_title('Fraction \"high frequency\" somatic mutations')\n",
    "fig.savefig('figs/frac_high_freq_somatic_mutations_mut_count.png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Somatic mutation rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# count the mutations and total coverage of each genotype (across all replicates)\n",
    "gen_counts = {g:[] for g in genotypes}\n",
    "gen_coverage = {g:[] for g in genotypes}\n",
    "for s in sample_names:\n",
    "    # add mutation count and coverage to the running totals for the genotype\n",
    "    gen_counts[s_to_g(s)].append(len(dfs_somatic[s]))\n",
    "    gen_coverage[s_to_g(s)].append(sample_call_cov[s])\n",
    "gen_count_sums = {g:sum(gen_counts[g]) for g in genotypes}\n",
    "gen_coverage_sums = {g:sum(gen_coverage[g]) for g in genotypes}\n",
    "gen_rates = {g:gen_count_sums[g] / gen_coverage_sums[g] for g in genotypes}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "{g:sum(gen_counts[g]) for g in genotypes}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do a t-test for each genotype compared to untreated WT\n",
    "p_ttest = dict()\n",
    "col_rates = [gen_counts['untreated'][i] / gen_coverage['untreated'][i] for i in range(len(gen_coverage['untreated']))]\n",
    "for g in genotypes:\n",
    "    if len(gen_coverage[g]) == 1: # you can't do a t-test with one replicate\n",
    "        p_ttest[g] = 1\n",
    "        continue\n",
    "    p_ttest[g] = stats.ttest_ind(col_rates, [gen_counts[g][i] / gen_coverage[g][i] for i in range(len(gen_coverage[g]))], equal_var=False)[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get 95% confidence intervals assuming t-distribution\n",
    "t_yerr = [dict(), dict()]\n",
    "for g in genotypes:\n",
    "    if len(gen_coverage[g]) > 1:\n",
    "        sem = np.std([gen_counts[g][i] / gen_coverage[g][i] for i in range(len(gen_coverage[g]))]) / (len(gen_coverage[g]) ** (1/2))\n",
    "        \n",
    "        low, high = stats.t.interval(0.95, len(gen_coverage[g]) - 1, loc=gen_rates[g], scale=sem)\n",
    "    else:\n",
    "        print(f'WARNING: no replicates for genotype {g}, using Poisson to calculate error bars')\n",
    "        low, high = stats.poisson(gen_count_sums[g]).interval(0.95)\n",
    "        low = low / gen_coverage_sums[g]\n",
    "        high = high / gen_coverage_sums[g]\n",
    "    t_yerr[0][g] = gen_rates[g] - low\n",
    "    t_yerr[1][g] = high - gen_rates[g]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sems = dict()\n",
    "for g in genotypes:\n",
    "    sem = np.std([gen_counts[g][i] / gen_coverage[g][i] for i in range(len(gen_coverage[g]))]) / (len(gen_coverage[g]) ** (1/2))\n",
    "    sems[g] = sem\n",
    "\n",
    "vis_genotypes = [g for g in genotypes if 'Ler-0' not in g and '1001' not in g]\n",
    "vis_samples = [s for s in sample_names if s_to_g(s) in vis_genotypes]\n",
    "\n",
    "cor_p_ttest = {g:p_ttest[g] * (len(vis_genotypes) - 1) for g in vis_genotypes}\n",
    "\n",
    "max_rate = 1.5 * 10**-6\n",
    "fig, axs = plt.subplots(2, figsize=(len(vis_genotypes), 6), sharex=True, gridspec_kw={'height_ratios':(1, 3)})\n",
    "axs[0].bar(range(len(vis_genotypes)), [gen_count_sums[g] for g in vis_genotypes], color='tab:orange')\n",
    "y = [min(max_rate, gen_count_sums[g] / gen_coverage_sums[g]) for g in vis_genotypes]\n",
    "# yerr = [[t_yerr[0][g] for g in vis_genotypes], [t_yerr[1][g] for g in vis_genotypes]]\n",
    "yerr = [sems[g] for g in vis_genotypes]\n",
    "bars = axs[1].bar(range(len(vis_genotypes)), y, yerr=yerr, color=[gen_cmap(g) for g in vis_genotypes], alpha=0.6)\n",
    "for i, g in enumerate(vis_genotypes):\n",
    "    axs[1].scatter([i + random.random() * 0.5 - 0.25 for x in gen_counts[g]], [min(max_rate, gen_counts[g][j] / gen_coverage[g][j]) for j in range(len(gen_counts[g]))], color=gen_cmap(g), edgecolors='none')\n",
    "    text = f'p<e{math.ceil(math.log10(max(10 ** -50, cor_p_ttest[g])))}' if cor_p_ttest[g] < 10 ** -3 else f'\\np={cor_p_ttest[g]:.3f}'\n",
    "    # axs[1].text(i, min(bars[i].get_height() * 1.1, max_rate), text, ha='center', fontsize=12)\n",
    "axs[1].set_xticks(range(len(vis_genotypes)))\n",
    "axs[1].set_xticklabels(vis_genotypes, rotation=45, fontsize=14, ha='right')\n",
    "axs[0].set_title('Somatic mutation counts by genotype', y=1)\n",
    "axs[1].set_title('Somatic mutation rates by genotype', y=1)\n",
    "axs[1].set_ylabel('Somatic mutations / callable bp')\n",
    "# axs[1].set_yscale('log')\n",
    "# axs[1].set_ylim(axs[1].get_ylim()[0], axs[1].get_ylim()[1] * 2.5)\n",
    "# axs[1].set_ylim(0, 6 * 10**-7)\n",
    "xlim = axs[1].get_xlim()\n",
    "axs[1].hlines(bars[0].get_height(), -1, len(gen_counts) + 1, color='black')\n",
    "# axs[1].hlines(8.25 * 10 ** -9, -1, len(gen_counts) + 1, color='tab:red')\n",
    "axs[1].set_xlim(xlim)\n",
    "fig.savefig('figs/somatic_mutation_rates_by_genotype.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "for i in range(len(rates)):\n",
    "    ax.plot(target_muts, p_power[i])\n",
    "xlim = ax.get_xlim()\n",
    "ax.hlines(0.05, -100, 10000, color='black')\n",
    "ax.hlines(0.0025, -100, 10000, color='black')\n",
    "ax.set_xlim(xlim)\n",
    "ax.legend(rates, bbox_to_anchor=(1, 1), title='Mutation rate\\nincrease over\\nCol-0')\n",
    "ax.set_yscale('log')\n",
    "ax.set_ylim(0.00001, 0.1)\n",
    "ax.set_title('Mutations needed to detect an increase over Col-0')\n",
    "ax.set_xlabel('Targeted number of mutations')\n",
    "ax.set_ylabel('Chi squared P-value (power 0.9)')\n",
    "fig.savefig('figs/mutations_needed_to_detect_an_increase_over_col-0.png', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_wt = pd.concat(dfs_somatic[:10])\n",
    "df_ner = pd.concat(dfs_somatic[10:])\n",
    "len(df_wt), len(df_ner)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = gtools.load_genome(genome_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# turn the mutation counts into a bedgraph\n",
    "with open('tmp/wt_muts.bg', 'w') as f:\n",
    "    for chrom in genome:\n",
    "        positions = list(df_wt[df_wt.chrom == chrom].pos)\n",
    "        unique_pos = sorted(list(set(positions)))\n",
    "        last_pos = -1\n",
    "        for pos in unique_pos:\n",
    "            if pos != 0: # avoids writing a [0, 0) block if there's a mutation at position 0 \n",
    "                f.write(f'{chrom}\\t{last_pos + 1}\\t{pos}\\t0\\n')\n",
    "            f.write(f'{chrom}\\t{pos}\\t{pos + 1}\\t{positions.count(pos)}\\n') # this won't merge two mutations at adjacent positions into one block\n",
    "            last_pos = pos\n",
    "        if last_pos + 1 != len(genome[chrom]): # avoids writing a [chrom len, chrom len) block if there's a mutation at the last base\n",
    "            f.write(f'{chrom}\\t{last_pos + 1}\\t{len(genome[chrom])}\\t0\\n')\n",
    "\n",
    "with open('tmp/ner_muts.bg', 'w') as f:\n",
    "    for chrom in genome:\n",
    "        positions = list(df_ner[df_ner.chrom == chrom].pos)\n",
    "        unique_pos = sorted(list(set(positions)))\n",
    "        last_pos = -1\n",
    "        for pos in unique_pos:\n",
    "            if pos != 0:\n",
    "                f.write(f'{chrom}\\t{last_pos + 1}\\t{pos}\\t0\\n')\n",
    "            f.write(f'{chrom}\\t{pos}\\t{pos + 1}\\t{positions.count(pos)}\\n') # this won't merge two mutations at adjacent positions into one block\n",
    "            last_pos = pos\n",
    "        if last_pos + 1 != len(genome[chrom]):\n",
    "            f.write(f'{chrom}\\t{last_pos + 1}\\t{len(genome[chrom])}\\t0\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make callable coverage arrays summing over wt or ner mutant samples\n",
    "wt_cov = None\n",
    "for d in list(call_cov.values())[:10]:\n",
    "    if wt_cov is None:\n",
    "        wt_cov = d.copy()\n",
    "    else:\n",
    "        for chrom in d:\n",
    "            wt_cov[chrom] += d[chrom]\n",
    "\n",
    "ner_cov = None\n",
    "for d in list(call_cov.values())[10:]:\n",
    "    if ner_cov is None:\n",
    "        ner_cov = d.copy()\n",
    "    else:\n",
    "        for chrom in d:\n",
    "            ner_cov[chrom] += d[chrom]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# turn the callable coverage arrays into bedgraphs\n",
    "with open('tmp/wt_cov.bg', 'w') as f:\n",
    "    for chrom in wt_cov:\n",
    "        start = 0\n",
    "        value = 0\n",
    "        for i, cov in tqdm(enumerate(wt_cov[chrom]), total=len(wt_cov[chrom])):\n",
    "            if cov != value:\n",
    "                if i != 0:\n",
    "                    f.write(f'{chrom}\\t{start}\\t{i}\\t{value}\\n')\n",
    "                start = i\n",
    "                value = cov\n",
    "        f.write(f'{chrom}\\t{start}\\t{len(genome[chrom])}\\t{value}\\n')\n",
    "    f.write(f'ChrC\\t0\\t154478\\t0\\n')\n",
    "    f.write(f'ChrM\\t0\\t367808\\t0\\n')\n",
    "        \n",
    "with open('tmp/ner_cov.bg', 'w') as f:\n",
    "    for chrom in ner_cov:\n",
    "        start = 0\n",
    "        value = 0\n",
    "        for i, cov in tqdm(enumerate(ner_cov[chrom]), total=len(ner_cov[chrom])):\n",
    "            if cov != value:\n",
    "                if i != -1:\n",
    "                    f.write(f'{chrom}\\t{start}\\t{i}\\t{value}\\n')\n",
    "                start = i\n",
    "                value = cov\n",
    "        f.write(f'{chrom}\\t{start}\\t{len(genome[chrom])}\\t{value}\\n')\n",
    "    f.write(f'ChrC\\t0\\t154478\\t0\\n')\n",
    "    f.write(f'ChrM\\t0\\t367808\\t0\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Mutation spectra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find every possible mutation 3bp context e.g. ACAT means ACA->ATA\n",
    "contexts = []\n",
    "for mut in ['CA', 'CG', 'CT', 'TA', 'TC', 'TG']: # SNP type e.g. A->C\n",
    "    for bef in 'ACGT':\n",
    "        for aft in 'ACGT':\n",
    "            contexts.append(f'{bef}{mut[0]}{aft}{mut[1]}')\n",
    "contexts.append('del')\n",
    "contexts.append('ins')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reverse complement a sequence\n",
    "comp = {'A':'T', 'T':'A', 'C':'G', 'G':'C'}\n",
    "def rc(seq):\n",
    "    ret = ''\n",
    "    for nuc in reversed(seq):\n",
    "        ret += comp[nuc]\n",
    "    return ret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# restrict_region = None\n",
    "# restrict_region = region_arrays['exon']\n",
    "\n",
    "# generate a null mutation spectrum\n",
    "# load callable coverage of first sample\n",
    "call_cov_0 = dict()\n",
    "call_cov_0_chroms = dict()\n",
    "for chrom in chrom_sizes:\n",
    "    call_cov_0[chrom] = np.load(f'{cov_prefixes[\"untreated_1\"]}{chrom}.npy')\n",
    "    call_cov_0_chroms[chrom] = np.sum(call_cov_0[chrom])\n",
    "\n",
    "# randomly sample positions in the genome based on callable coverage\n",
    "sample_size = 2000000\n",
    "sampled_positions = dict()\n",
    "for chrom in tqdm(chrom_sizes):\n",
    "    k = int(sample_size * np.sum(call_cov_0[chrom]) / sum([np.sum(call_cov_0[c]) for c in chrom_sizes]))\n",
    "    sampled_positions[chrom] = random.choices(range(len(call_cov_0[chrom])), weights=call_cov_0[chrom], k=k)\n",
    "\n",
    "# build a null spectrum using the randomly sampled positions, won't include SNP info\n",
    "context_freq = {c[0:3]:0 for c in contexts}\n",
    "for chrom in tqdm(chrom_sizes):\n",
    "    for pos in sampled_positions[chrom]:\n",
    "        context = genome[chrom][pos - 1:pos + 2]\n",
    "        if context[0] not in 'ACGT' or context[1] not in 'ACGT' or context[2] not in 'ACGT': # skip context with N, W, R, Y...\n",
    "            continue\n",
    "        if context[1] == 'A' or context[1] == 'G': # reverse complement contexts centered on G or T\n",
    "            context = rc(context)\n",
    "        context_freq[context] += 1\n",
    "context_freq = {c:context_freq[c] / sum(context_freq.values()) for c in context_freq} # so they sum to 1\n",
    "context_freq['ins'] = 1\n",
    "context_freq['del'] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate the spectrum for each sample\n",
    "spectra = {sample:{c:0 for c in contexts} for sample in sample_names}\n",
    "for sample in sample_names:\n",
    "    warned = False\n",
    "    for r in dfs_somatic[sample].itertuples():\n",
    "        if r.ref == '*':\n",
    "            spectra[sample]['ins'] += 1\n",
    "        elif r.alt == '*':\n",
    "            spectra[sample]['del'] += 1\n",
    "        else:\n",
    "            alt = r.alt\n",
    "            context = genome[r.chrom][r.pos - 1:r.pos + 2]\n",
    "            if not set(context).issubset({'A', 'C', 'G', 'T'}):\n",
    "                print(f'WARNING: context of {context} in sample {sample} for mutation {r}, ignoring this mutation')\n",
    "                continue\n",
    "            if context[1] != r.ref:\n",
    "                context = context[0] + r.ref + context[2]\n",
    "                if not warned:\n",
    "                    print(f'WARNING: ref base does not match reference genome in sample {sample} for mutation {r}, hiding further warnings for this sample')\n",
    "                    warned = True\n",
    "            if r.ref == 'A' or r.ref == 'G': # if the mut has the reference as G or T, reverse copmlement the context\n",
    "                context = rc(context)\n",
    "                alt = comp[r.alt]\n",
    "            spectra[sample][context + alt] += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# normalize by frequency of sequence context, total callable coverage, and avg genome-wide mutation rate, the value is now the mutation rate relative to genome-wide average\n",
    "# norm rate = num mutations in context / (frequency of context * total coverage) / (total mutations / total coverage)\n",
    "#                                                 ^coverage of context                 ^genome-wide avg mutation rate\n",
    "# norm rate = num mutations in context / frequency of context / total mutations\n",
    "def normalize_spectrum(spec, spec_freq, total_mutations):\n",
    "    s = dict()\n",
    "    for c in spec:\n",
    "        freq_c = c[:-1] if c != 'ins' and c != 'del' else c\n",
    "        s[c] = spec[c] / spec_freq[freq_c] / total_mutations\n",
    "    return s\n",
    "\n",
    "norm_spectra = dict()\n",
    "for sample, spec in spectra.items():\n",
    "    norm_spectra[sample] = normalize_spectrum(spec, context_freq, sum(spec.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ignore_samples = []\n",
    "# vis_samples = sample_names[:8]\n",
    "# vis_genotypes = list({s_to_g(s):0 for s in vis_samples})\n",
    "\n",
    "# mut_colors = ['tab:cyan'] * 16 + ['black'] * 16 + ['tab:red'] * 16 + ['tab:grey'] * 16 + ['tab:green'] * 16 + ['tab:pink'] * 16 + ['tab:purple'] + ['tab:brown']\n",
    "\n",
    "# fig, axs = plt.subplots(len(vis_samples), figsize=(14, 0.7 * len(vis_samples)), sharex=True, sharey=True)\n",
    "# for i, sample in enumerate(vis_samples):\n",
    "#     axs[i].bar(range(len(contexts)), norm_spectra[sample].values(), color=mut_colors)\n",
    "#     title = sample + ' (n=' + str(len(dfs_somatic[sample])) + ')'\n",
    "#     axs[i].set_title(title, y=0.5, x=0.01, loc='left')\n",
    "#     axs[i].ticklabel_format(style='plain')\n",
    "# axs[-1].set_xticks(range(len(contexts)))\n",
    "# axs[-1].set_xticklabels([c[:3] for c in spectra[vis_samples[0]].keys()], rotation=90, size=6)\n",
    "# for pos, label in zip([7.5 + 16 * i for i in range(6)], ['C -> A', 'C -> G', 'C -> T', 'T -> A', 'T -> C', 'T -> G']): # trying to set separate major and minor ticks has a weird bug\n",
    "#     axs[-1].text(pos, -1 * axs[-1].get_ylim()[1], label)\n",
    "# axs[len(axs) // 2].set_ylabel('Mutation rate relative to avg')\n",
    "# # axs[-1].set_ylim(0, 0.15)\n",
    "# fig.suptitle('Spectra of somatic mutations', y=0.9)\n",
    "# fig.savefig('figs/mutation_spectra_by_sample.png', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# group spectra from similar treatments and sum them for a new plot\n",
    "\n",
    "# sum the spectra of all samples for a given genotype\n",
    "grouped_spectra = {g:{c:0 for c in contexts} for g in genotypes}\n",
    "for sample in sample_names:\n",
    "    genotype = s_to_g(sample)\n",
    "    for c in contexts:\n",
    "        grouped_spectra[genotype][c] += spectra[sample][c]\n",
    "\n",
    "# make a 95% confidence interval for each context using biniomial distribution\n",
    "gs_lower = {g:dict() for g in genotypes}\n",
    "gs_upper = {g:dict() for g in genotypes}\n",
    "for genotype in genotypes:\n",
    "    for c in contexts:\n",
    "        total = sum(grouped_spectra[genotype].values())\n",
    "        lower, upper = stats.binom.interval(0.95, total, max(1, grouped_spectra[genotype][c]) / total)\n",
    "        gs_lower[genotype][c] = lower\n",
    "        gs_upper[genotype][c] = upper\n",
    "\n",
    "# normalize the genotype spectra and confidence intervals; convert confidence intervals to error amounts\n",
    "norm_grouped_spectra = dict()\n",
    "ngs_yerr_0 = dict()\n",
    "ngs_yerr_1 = dict()\n",
    "for genotype in genotypes:\n",
    "    norm_grouped_spectra[genotype] = normalize_spectrum(grouped_spectra[genotype], context_freq, sum(grouped_spectra[genotype].values()))\n",
    "    lower = normalize_spectrum(gs_lower[genotype], context_freq, sum(grouped_spectra[genotype].values()))\n",
    "    upper = normalize_spectrum(gs_upper[genotype], context_freq, sum(grouped_spectra[genotype].values()))\n",
    "    ngs_yerr_0[genotype] = {c:abs(norm_grouped_spectra[genotype][c] - lower[c]) for c in contexts}\n",
    "    ngs_yerr_1[genotype] = {c:abs(norm_grouped_spectra[genotype][c] - upper[c]) for c in contexts}\n",
    "\n",
    "\n",
    "# count the number of mutations in each genotype\n",
    "grouped_counts = {genotype: sum(grouped_spectra[genotype].values()) for genotype in genotypes}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_contexts = [c for c in contexts if c != 'ins' and c !='del']\n",
    "\n",
    "sems = {g:dict() for g in genotypes}\n",
    "for g in genotypes:\n",
    "    for context in full_contexts:\n",
    "        sems[g][context] = np.std([norm_spectra[s][context] for s in g_to_s(g)]) / (len(g_to_s(g)) ** (1/2))\n",
    "\n",
    "# ignore_samples = []\n",
    "vis_genotypes = genotypes\n",
    "vis_samples = [s for s in sample_names if s_to_g(s) in vis_genotypes]\n",
    "\n",
    "fig, axs = plt.subplots(len(vis_genotypes), figsize=(11, len(vis_genotypes) * 1.5), sharex=True, sharey=True)\n",
    "full_colors = ['tab:cyan'] * 16 + ['black'] * 16 + ['tab:red'] * 16 + ['tab:grey'] * 16 + ['tab:green'] * 16 + ['tab:pink'] * 16\n",
    "for i, genotype in enumerate(vis_genotypes):\n",
    "    samples = g_to_s(genotype)\n",
    "    # yerrs = [[ngs_yerr_0[genotype][c] for c in full_contexts], [ngs_yerr_1[genotype][c] for c in full_contexts]]\n",
    "    axs[i].bar(range(len(full_contexts)), [norm_grouped_spectra[genotype][c] for c in full_contexts], color=full_colors, alpha=0.6, yerr=sems[genotype].values())\n",
    "    # axs[i].errorbar([x + 0.1 for x in range(len(full_contexts))], [norm_grouped_spectra[genotype][c] for c in full_contexts], yerr=list(sems[genotype].values()), fmt='none')\n",
    "    title = genotype + ' (n=' + str(grouped_counts[genotype]) + ')'\n",
    "    if len(g_to_s(genotype)) > 1 and len(g_to_s(genotype)) < 30:\n",
    "        for sample in g_to_s(genotype):\n",
    "            size = 2 if len(g_to_s(genotype)) < 20 else 0.5\n",
    "            axs[i].scatter([x + random.random() * 0.4 - 0.2 for x in range(len(full_contexts))], [norm_spectra[sample][c] for c in full_contexts], color=full_colors, s=size, edgecolors='none')\n",
    "    axs[i].set_title(title, y=0.55, x=0.01, loc='left')\n",
    "    axs[i].ticklabel_format(style='plain')\n",
    "\n",
    "axs[-1].set_xticks(range(len(full_contexts)))\n",
    "axs[-1].set_xticklabels([c[:3] for c in full_contexts], rotation=90, size=6)\n",
    "axs[-1].set_ylim(0, 5.8)\n",
    "for pos, label in zip([7.5 + 16 * i for i in range(6)], ['C -> A', 'C -> G', 'C -> T', 'T -> A', 'T -> C', 'T -> G']): # trying to set separate major and minor ticks has a weird bug\n",
    "    axs[-1].text(pos, -1 * axs[-1].get_ylim()[1], label)\n",
    "axs[len(axs) // 2].set_ylabel('mutation/polymorphism rate relative to uniform')\n",
    "fig.suptitle('Grouped spectra of somatic mutations', y=0.92)\n",
    "fig.savefig('figs/mutation_spectra_by_sample_groups.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Simple spectra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make the above plot, but simplify to just type of SNP (no 3bp context)\n",
    "simple_contexts = ['CA', 'CG', 'CT', 'TA', 'TC', 'TG', 'del', 'ins']\n",
    "\n",
    "# make a simple version of context_freq\n",
    "simple_context_freq = {'C':0, 'T':0, 'del':0, 'ins':0}\n",
    "for c in context_freq:\n",
    "    simple = c if c == 'ins' or c == 'del' else c[1]\n",
    "    simple_context_freq[simple] += context_freq[c]\n",
    "\n",
    "# make simple versions of the genotype spectra\n",
    "simple_spectra = {g:{c:0 for c in simple_contexts} for g in genotypes}\n",
    "simple_spectra_samples = {s:{c:0 for c in simple_contexts} for s in sample_names}\n",
    "for sample in sample_names:\n",
    "    genotype = s_to_g(sample)\n",
    "    for c in contexts:\n",
    "        simple_c = c if c == 'ins' or c == 'del' else c[1] + c[3]\n",
    "        simple_spectra[genotype][simple_c] += spectra[sample][c]\n",
    "        simple_spectra_samples[sample][simple_c] += spectra[sample][c]\n",
    "\n",
    "# make 95% confidence intervals of the simple spectra\n",
    "ss_lower = {g:dict() for g in genotypes}\n",
    "ss_upper = {g:dict() for g in genotypes}\n",
    "for genotype in genotypes:\n",
    "    for c in simple_contexts:\n",
    "        total = sum(simple_spectra[genotype].values())\n",
    "        lower, upper = stats.binom.interval(0.95, total, max(1, simple_spectra[genotype][c]) / total)\n",
    "        ss_lower[genotype][c] = lower\n",
    "        ss_upper[genotype][c] = upper\n",
    "\n",
    "# normalize spectra and confidence intervals, convert confidence intervals to error amounts\n",
    "norm_simple_spectra = dict()\n",
    "norm_simple_spectra_samples = dict()\n",
    "nss_yerr_0 = dict()\n",
    "nss_yerr_1 = dict()\n",
    "for genotype in genotypes:\n",
    "    norm_simple_spectra[genotype] = normalize_spectrum(simple_spectra[genotype], simple_context_freq, sum(simple_spectra[genotype].values()))\n",
    "    lower = normalize_spectrum(ss_lower[genotype], simple_context_freq, sum(simple_spectra[genotype].values()))\n",
    "    upper = normalize_spectrum(ss_upper[genotype], simple_context_freq, sum(simple_spectra[genotype].values()))\n",
    "    nss_yerr_0[genotype] = {c:abs(norm_simple_spectra[genotype][c] - lower[c]) for c in simple_contexts}\n",
    "    nss_yerr_1[genotype] = {c:abs(norm_simple_spectra[genotype][c] - upper[c]) for c in simple_contexts}\n",
    "for sample in sample_names:\n",
    "    norm_simple_spectra_samples[sample] = normalize_spectrum(simple_spectra_samples[sample], simple_context_freq, sum(simple_spectra_samples[sample].values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sems = {g:dict() for g in genotypes}\n",
    "for g in genotypes:\n",
    "    for context in simple_contexts:\n",
    "        sems[g][context] = np.std([norm_simple_spectra_samples[s][context] for s in g_to_s(g)]) / (len(g_to_s(g)) ** (1/2))\n",
    "\n",
    "fig, axs = plt.subplots(len(vis_genotypes), figsize=(4, len(vis_genotypes) * 1.5), sharex=True, sharey=True)\n",
    "simple_colors = ['tab:cyan', 'black', 'tab:red', 'tab:grey', 'tab:green', 'tab:pink', 'tab:brown', 'tab:purple']\n",
    "for i, genotype in enumerate(vis_genotypes):\n",
    "    samples = g_to_s(genotype)\n",
    "    yerrs = [list(nss_yerr_0[genotype].values()), list(nss_yerr_1[genotype].values())]\n",
    "    axs[i].bar(range(len(simple_contexts)), norm_simple_spectra[genotype].values(), color=simple_colors, alpha=0.6, yerr=sems[genotype].values())\n",
    "    title = genotype + ' (n=' + str(grouped_counts[genotype]) + ')'\n",
    "    for sample in g_to_s(genotype):\n",
    "        size = 8 if len(g_to_s(genotype)) < 20 else 1\n",
    "        axs[i].scatter([x + random.random() * 0.5 - 0.25 for x in range(len(simple_contexts))], norm_simple_spectra_samples[sample].values(), color=simple_colors, s=size, edgecolors='none')\n",
    "    axs[i].set_title(title, y=0.55, x=0.01, loc='left')\n",
    "    axs[i].ticklabel_format(style='plain')\n",
    "\n",
    "axs[-1].set_xticks(range(len(simple_contexts)))\n",
    "axs[-1].set_xticklabels([c if c == 'ins' or c == 'del' else f'{c[0]}>{c[1]}' for c in simple_contexts], size=8)\n",
    "axs[len(axs) // 2].set_ylabel('relative substitution rate')\n",
    "axs[-1].set_ylim(0, axs[-1].get_ylim()[1])\n",
    "fig.suptitle('Simple substitution spectra', y=0.92)\n",
    "fig.savefig('figs/simple_mutation_spectra_by_sample_groups.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for g in vis_genotypes:\n",
    "    for context in simple_contexts:\n",
    "        p = stats.ttest_ind([norm_simple_spectra_samples[s][context] for s in g_to_s(g)], [norm_simple_spectra_samples[s][context] for s in g_to_s('untreated')], equal_var=True).pvalue\n",
    "        print(g, context, p * len(simple_contexts) * (len(vis_genotypes) - 1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CG and YC spectra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make the same plot as above, but for (C>T)G and (C>T)H sites\n",
    "\n",
    "cg_contexts = ['NCGT', 'NCHT']\n",
    "\n",
    "# get frequency of CG and CH sites\n",
    "cg_freq = {'NCG':0, 'NCH':0}\n",
    "for c in context_freq:\n",
    "    if c == 'ins' or c == 'del':\n",
    "        continue\n",
    "    if c[1] == 'C' and c[2] == 'G':\n",
    "        cg_freq['NCG'] += context_freq[c]\n",
    "    elif c[1] == 'C':\n",
    "        cg_freq['NCH'] += context_freq[c]\n",
    "\n",
    "# get the counts of CG and CH mutations\n",
    "cg_spectra = {g:{c:0 for c in cg_contexts} for g in genotypes}\n",
    "cg_spectra_samples = {s:{c:0 for c in cg_contexts} for s in sample_names}\n",
    "for sample in sample_names:\n",
    "    genotype = s_to_g(sample)\n",
    "    for c in contexts:\n",
    "        if c[1] == 'C' and c[2] == 'G' and c[3] == 'T':\n",
    "            cg_spectra[genotype]['NCGT'] += spectra[sample][c]\n",
    "            cg_spectra_samples[sample]['NCGT'] += spectra[sample][c]\n",
    "        elif c[1] == 'C' and c[3] == 'T':\n",
    "            cg_spectra[genotype]['NCHT'] += spectra[sample][c]\n",
    "            cg_spectra_samples[sample]['NCHT'] += spectra[sample][c]\n",
    "\n",
    "# make 95% confidence intervals\n",
    "cg_lower = {g:dict() for g in genotypes}\n",
    "cg_upper = {g:dict() for g in genotypes}\n",
    "for genotype in genotypes:\n",
    "    for c in cg_contexts:\n",
    "        total = sum(cg_spectra[genotype].values())\n",
    "        lower, upper = stats.binom.interval(0.95, total, max(1, cg_spectra[genotype][c]) / total)\n",
    "        cg_lower[genotype][c] = lower\n",
    "        cg_upper[genotype][c] = upper\n",
    "\n",
    "# normalize spectra and confidence intervals, convert confidence intervals to error amounts\n",
    "norm_cg_spectra = dict()\n",
    "norm_cg_spectra_samples = dict()\n",
    "ncg_yerr_0 = dict()\n",
    "ncg_yerr_1 = dict()\n",
    "for genotype in genotypes:\n",
    "    norm_cg_spectra[genotype] = normalize_spectrum(cg_spectra[genotype], cg_freq, sum(cg_spectra[genotype].values()))\n",
    "    lower = normalize_spectrum(cg_lower[genotype], cg_freq, sum(cg_spectra[genotype].values()))\n",
    "    upper = normalize_spectrum(cg_upper[genotype], cg_freq, sum(cg_spectra[genotype].values()))\n",
    "    ncg_yerr_0[genotype] = {c:abs(norm_cg_spectra[genotype][c] - lower[c]) for c in cg_contexts}\n",
    "    ncg_yerr_1[genotype] = {c:abs(norm_cg_spectra[genotype][c] - upper[c]) for c in cg_contexts}\n",
    "for sample in sample_names:\n",
    "    norm_cg_spectra_samples[sample] = normalize_spectrum(cg_spectra_samples[sample], cg_freq, sum(cg_spectra_samples[sample].values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sems = {g:dict() for g in genotypes}\n",
    "for g in genotypes:\n",
    "    for context in cg_contexts:\n",
    "        sems[g][context] = np.std([norm_cg_spectra_samples[s][context] for s in g_to_s(g)]) / (len(g_to_s(g)) ** (1/2))\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(len(vis_genotypes), 3))\n",
    "\n",
    "c = gtools.light_cmap[0]\n",
    "samples = g_to_s(genotype)\n",
    "# yerrs = [[ncg_yerr_0[g]['NCGT'] for g in vis_genotypes], [ncg_yerr_1[g]['NCGT'] for g in vis_genotypes]]\n",
    "yerrs = [sems[g]['NCGT'] for g in vis_genotypes]\n",
    "bar_cg = ax.bar(range(0, len(vis_genotypes) * 3, 3), [norm_cg_spectra[genotype]['NCGT'] for genotype in vis_genotypes], yerr=yerrs, color=c, alpha=0.6)\n",
    "ax.scatter([vis_genotypes.index(s_to_g(s)) * 3 + random.random() * 0.5 - 0.25 for s in vis_samples], [norm_cg_spectra_samples[s]['NCGT'] for s in vis_samples], color=c, s=[6 if len(g_to_s(s_to_g(s))) < 20 else 0.5 for s in vis_samples])\n",
    "\n",
    "c = gtools.dark_cmap[0]\n",
    "# yerrs = [[ncg_yerr_0[g]['NCHT'] for g in vis_genotypes], [ncg_yerr_1[g]['NCHT'] for g in vis_genotypes]]\n",
    "yerrs = [sems[g]['NCHT'] for g in vis_genotypes]\n",
    "bar_ch = ax.bar(range(1, len(vis_genotypes) * 3 + 1, 3), [norm_cg_spectra[genotype]['NCHT'] for genotype in vis_genotypes], yerr=yerrs, color=c, alpha=0.6)\n",
    "ax.scatter([vis_genotypes.index(s_to_g(s)) * 3 + 1 + random.random() * 0.5 - 0.25 for s in vis_samples], [norm_cg_spectra_samples[s]['NCHT'] for s in vis_samples], color=c, s=[6 if len(g_to_s(s_to_g(s))) < 20 else 0.5 for s in vis_samples])\n",
    "\n",
    "ax.set_xticks([x + 0.5 for x in range(0, len(vis_genotypes) * 3, 3)], labels=vis_genotypes)\n",
    "ax.legend([bar_cg, bar_ch], ['(C>T)G', '(C>T)H'])\n",
    "ax.set_ylim((0, 11.2))\n",
    "fig.suptitle('CG site vs CH site C>T rate', y=0.95)\n",
    "fig.savefig('figs/cg_vs_ch_spectrum.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for g in vis_genotypes:\n",
    "    cg = [norm_cg_spectra_samples[s]['NCGT'] for s in g_to_s(g)]\n",
    "    ch = [norm_cg_spectra_samples[s]['NCHT'] for s in g_to_s(g)]\n",
    "    p = stats.ttest_ind(cg, ch).pvalue\n",
    "    print(g, norm_cg_spectra[g]['NCGT'] / norm_cg_spectra[g]['NCHT'], p * len(vis_genotypes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make the same plot as above, but for Y(C>T) and R(C>T) sites\n",
    "\n",
    "yc_contexts = ['YCNT', 'RCNT']\n",
    "\n",
    "# get frequency of YC and RC sites\n",
    "yc_freq = {'YCN':0, 'RCN':0}\n",
    "for c in context_freq:\n",
    "    if c == 'ins' or c == 'del':\n",
    "        continue\n",
    "    if (c[0] == 'C' or c[0] == 'T') and c[1] == 'C':\n",
    "        yc_freq['YCN'] += context_freq[c]\n",
    "    elif c[1] == 'C':\n",
    "        yc_freq['RCN'] += context_freq[c]\n",
    "\n",
    "# get the counts of YC and RC mutations\n",
    "yc_spectra = {g:{c:0 for c in yc_contexts} for g in genotypes}\n",
    "yc_spectra_samples = {s:{c:0 for c in yc_contexts} for s in sample_names}\n",
    "for sample in sample_names:\n",
    "    genotype = s_to_g(sample)\n",
    "    for c in contexts:\n",
    "        if (c[0] == 'C' or c[0] == 'T') and c[1] == 'C' and c[3] == 'T':\n",
    "            yc_spectra[genotype]['YCNT'] += spectra[sample][c]\n",
    "            yc_spectra_samples[sample]['YCNT'] += spectra[sample][c]\n",
    "        elif c[1] == 'C' and c[3] == 'T':\n",
    "            yc_spectra[genotype]['RCNT'] += spectra[sample][c]\n",
    "            yc_spectra_samples[sample]['RCNT'] += spectra[sample][c]\n",
    "\n",
    "# make 95% confidence intervals\n",
    "yc_lower = {g:dict() for g in genotypes}\n",
    "yc_upper = {g:dict() for g in genotypes}\n",
    "for genotype in genotypes:\n",
    "    for c in yc_contexts:\n",
    "        total = sum(yc_spectra[genotype].values())\n",
    "        lower, upper = stats.binom.interval(0.95, total, max(1, yc_spectra[genotype][c]) / total)\n",
    "        yc_lower[genotype][c] = lower\n",
    "        yc_upper[genotype][c] = upper\n",
    "\n",
    "# normalize spectra and confidence intervals, convert confidence intervals to error amounts\n",
    "norm_yc_spectra = dict()\n",
    "norm_yc_spectra_samples = dict()\n",
    "nyc_yerr_0 = dict()\n",
    "nyc_yerr_1 = dict()\n",
    "for genotype in genotypes:\n",
    "    norm_yc_spectra[genotype] = normalize_spectrum(yc_spectra[genotype], yc_freq, sum(yc_spectra[genotype].values()))\n",
    "    lower = normalize_spectrum(yc_lower[genotype], yc_freq, sum(yc_spectra[genotype].values()))\n",
    "    upper = normalize_spectrum(yc_upper[genotype], yc_freq, sum(yc_spectra[genotype].values()))\n",
    "    nyc_yerr_0[genotype] = {c:abs(norm_yc_spectra[genotype][c] - lower[c]) for c in yc_contexts}\n",
    "    nyc_yerr_1[genotype] = {c:abs(norm_yc_spectra[genotype][c] - upper[c]) for c in yc_contexts}\n",
    "for sample in sample_names:\n",
    "    norm_yc_spectra_samples[sample] = normalize_spectrum(yc_spectra_samples[sample], yc_freq, sum(yc_spectra_samples[sample].values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sems = {g:dict() for g in genotypes}\n",
    "for g in genotypes:\n",
    "    for context in yc_contexts:\n",
    "        sems[g][context] = np.std([norm_yc_spectra_samples[s][context] for s in g_to_s(g)]) / (len(g_to_s(g)) ** (1/2))\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(len(vis_genotypes), 3))\n",
    "\n",
    "c = gtools.light_cmap[9]\n",
    "# yerrs = [[nyc_yerr_0[g]['YCNT'] for g in vis_genotypes], [nyc_yerr_1[g]['YCNT'] for g in vis_genotypes]]\n",
    "yerrs = [sems[g]['YCNT'] for g in vis_genotypes]\n",
    "bar_yc = ax.bar(range(0, len(vis_genotypes) * 3, 3), [norm_yc_spectra[genotype]['YCNT'] for genotype in vis_genotypes], yerr=yerrs, color=c, alpha=0.6)\n",
    "ax.scatter([vis_genotypes.index(s_to_g(s)) * 3 + random.random() * 0.5 - 0.25 for s in vis_samples], [norm_yc_spectra_samples[s]['YCNT'] for s in vis_samples], color=c, s=[6 if len(g_to_s(s_to_g(s))) < 20 else 0.5 for s in vis_samples])\n",
    "\n",
    "c = gtools.dark_cmap[9]\n",
    "# yerrs = [[nyc_yerr_0[g]['RCNT'] for g in vis_genotypes], [nyc_yerr_1[g]['RCNT'] for g in vis_genotypes]]\n",
    "yerrs = [sems[g]['RCNT'] for g in vis_genotypes]\n",
    "bar_rc = ax.bar(range(1, len(vis_genotypes) * 3 + 1, 3), [norm_yc_spectra[genotype]['RCNT'] for genotype in vis_genotypes], yerr=yerrs, color=c, alpha=0.6)\n",
    "ax.scatter([vis_genotypes.index(s_to_g(s)) * 3 + 1 + random.random() * 0.5 - 0.25 for s in vis_samples], [norm_yc_spectra_samples[s]['RCNT'] for s in vis_samples], color=c, s=[6 if len(g_to_s(s_to_g(s))) < 20 else 0.5 for s in vis_samples])\n",
    "    \n",
    "ax.set_xticks([x + 0.5 for x in range(0, len(vis_genotypes) * 3, 3)], labels=vis_genotypes)\n",
    "ax.legend([bar_yc, bar_rc], ['Y(C>T)', 'R(C>T)'])\n",
    "fig.suptitle('YC site vs RC site C>T rate', y=0.95)\n",
    "fig.savefig('figs/yc_vs_rc_spectrum.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for g in vis_genotypes:\n",
    "    yc = [norm_yc_spectra_samples[s]['YCNT'] for s in g_to_s(g)]\n",
    "    rc = [norm_yc_spectra_samples[s]['RCNT'] for s in g_to_s(g)]\n",
    "    p = stats.ttest_ind(yc, rc).pvalue\n",
    "    print(g, norm_yc_spectra[g]['YCNT'] / norm_yc_spectra[g]['RCNT'], p * len(vis_genotypes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "col_diff = []\n",
    "for s in g_to_s('untreated'):\n",
    "    col_diff.append(norm_yc_spectra_samples[s]['YCNT'] / norm_yc_spectra_samples[s]['RCNT'])\n",
    "stats.ttest_1samp(col_diff, norm_yc_spectra['MA lines']['YCNT'] / norm_yc_spectra['MA lines']['RCNT'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Indels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "indels = {g:{x:0 for x in range(-11, 12)} for g in genotypes}\n",
    "for s in sample_names:\n",
    "    for r in dfs_somatic[s].itertuples():\n",
    "        if r.ref == '*':\n",
    "            indels[s_to_g(s)][min(len(r.alt), 11)] += 1\n",
    "        elif r.alt == '*':\n",
    "            indels[s_to_g(s)][-1 * min(len(r.ref), 11)] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(len(vis_genotypes), figsize=(6, len(vis_genotypes) * 1), sharex=True)\n",
    "for i, g in enumerate(vis_genotypes):\n",
    "    axs[i].bar(indels[g].keys(), indels[g].values(), color=['tab:brown'] * 11 + ['tab:purple'] * 12, alpha=0.6)\n",
    "    axs[i].set_title(g, x=0.01, y=0.5, ha='left')\n",
    "axs[-1].set_xlabel('deletion (bp)                   insertion (bp)')\n",
    "axs[-1].set_xticks([-11, -5, -1, 1, 5, 11], labels=['>10', '5', '1', '1', '5', '>10'])\n",
    "axs[len(axs) // 2].set_ylabel('count')\n",
    "fig.suptitle('indel sizes')\n",
    "fig.savefig('figs/indel_sizes.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a correlation heatmap using all variant contexts\n",
    "df_spectra = pd.DataFrame(norm_grouped_spectra)\n",
    "df_spectra = df_spectra.transpose().rename(columns={i:genotypes[i] for i in range(len(genotypes))})\n",
    "df_spectra['1001'] = list(norm_spectra[-1].values())\n",
    "\n",
    "\n",
    "fig, axs = plt.subplots(2, figsize=(6, 9), gridspec_kw={'height_ratios': [1, 4], 'hspace':0})\n",
    "Z = cluster.hierarchy.linkage(df_spectra.transpose(), method='ward', optimal_ordering=True, metric='euclidean')\n",
    "dn = cluster.hierarchy.dendrogram(Z, ax=axs[0], no_labels=True)\n",
    "df_spectra = df_spectra[[genotypes[i] if i < len(genotypes) else '1001' for i in dn['leaves']]]\n",
    "df_corr = df_spectra.corr('pearson')\n",
    "axs[1].imshow(df_corr)\n",
    "axs[0].axis('off')\n",
    "# axs[1].xaxis.tick_top()\n",
    "axs[1].set_xticks(range(len(dn['leaves'])))\n",
    "axs[1].set_xticklabels([genotypes[i] if i < len(genotypes) else '1001' for i in dn['leaves']], rotation=90)\n",
    "axs[1].set_yticks(range(len(dn['leaves'])))\n",
    "axs[1].set_yticklabels([genotypes[i] if i < len(genotypes) else '1001' for i in dn['leaves']])\n",
    "fig.savefig('figs/mutation_spectra_heatmap.png', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find variant contexts with high variance for feature selection\n",
    "variances = []\n",
    "high_var_contexts = []\n",
    "for c in contexts:\n",
    "    rates = [spec[c] for spec in norm_grouped_spectra]\n",
    "    var = np.var(rates)\n",
    "    variances.append(var)\n",
    "    if var > 0.00005:\n",
    "        high_var_contexts.append(c)\n",
    "        print(c, np.var(rates))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(variances, bins=[x / 100000 for x in range(100)], cumulative=1, histtype=\"step\")\n",
    "plt.show()\n",
    "# 0.00005"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check the variance of each context\n",
    "# fig, axs = plt.subplots(len(genotypes) + 2, figsize=(14, len(genotypes)), sharex=True)\n",
    "# colors = ['tab:blue'] * 16 + ['tab:orange'] * 16 + ['tab:green'] * 16 + ['tab:red'] * 16 + ['tab:purple'] * 16 + ['tab:cyan'] * 16 + ['tab:brown', 'tab:grey']\n",
    "# for i in range(len(genotypes)):\n",
    "#     axs[i].bar(range(len(contexts)), norm_grouped_spectra[i].values(), color=colors, alpha=0.6)\n",
    "#     title = genotypes[i] + ' (n=' + str(grouped_counts[i]) + ')'\n",
    "#     for j in range(len(sample_names)):\n",
    "#         if sample_names[j].rsplit('-', 1)[0] == genotypes[i]:\n",
    "#             axs[i].scatter([x + random.random() * 0.5 - 0.25 for x in range(len(contexts))], norm_spectra[j].values(), color=colors, s=1)\n",
    "#     axs[i].set_title(title, y=0.55, x=0.01, loc='left')\n",
    "#     axs[i].ticklabel_format(style='plain')\n",
    "# axs[-2].bar(range(len(contexts)), norm_spectra[-1].values(), color=colors, alpha=0.6)\n",
    "# axs[-2].set_title(f'1001 genomes (n={len(df_1001)})', y=0.55, x=0.01, loc='left')\n",
    "\n",
    "# axs[-1].bar(range(len(contexts)), variances)\n",
    "\n",
    "# axs[-1].ticklabel_format(style='plain')\n",
    "# axs[-1].set_xticks(range(len(contexts)))\n",
    "# axs[-1].set_xticklabels([c[0] + c[2] + c[-1] if c != 'ins' and c != 'del' else c for c in spectra[0].keys()], rotation=90, size=6)\n",
    "# for pos, label in zip([7.5 + 16 * i for i in range(6)], ['A -> C', 'A -> G', 'A -> T', 'C -> A', 'C -> G', 'C -> T']): # trying to set separate major and minor ticks has a weird bug\n",
    "#     axs[-1].text(pos, -0.001, label)\n",
    "# axs[len(axs) // 2].set_ylabel('Proportion of mutations (context frequency normalized)')\n",
    "# for ax in axs[:-1]:\n",
    "#     ax.set_ylim(0, 0.15)\n",
    "# fig.suptitle('Grouped spectra of somatic mutations', y=0.92)\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make the same correlation heatmap, but only use the high variance contexts\n",
    "df_spectra = pd.DataFrame([{context:rate for context, rate in spec.items() if context in high_var_contexts} for spec in norm_grouped_spectra])\n",
    "df_spectra = df_spectra.transpose().rename(columns={i:genotypes[i] for i in range(len(genotypes))})\n",
    "df_spectra['1001'] = [rate for context, rate in norm_spectra[-1].items() if context in high_var_contexts]\n",
    "\n",
    "\n",
    "fig, axs = plt.subplots(2, figsize=(6, 9), gridspec_kw={'height_ratios': [1, 4], 'hspace':0})\n",
    "Z = cluster.hierarchy.linkage(df_spectra.transpose(), 'ward', optimal_ordering=True)\n",
    "dn = cluster.hierarchy.dendrogram(Z, ax=axs[0], no_labels=True)\n",
    "df_spectra = df_spectra[[genotypes[i] if i < len(genotypes) else '1001' for i in dn['leaves']]]\n",
    "df_corr = df_spectra.corr('pearson')\n",
    "axs[1].imshow(df_corr)\n",
    "axs[0].axis('off')\n",
    "# axs[1].xaxis.tick_top()\n",
    "axs[1].set_xticks(range(len(dn['leaves'])))\n",
    "axs[1].set_xticklabels([genotypes[i] if i < len(genotypes) else '1001' for i in dn['leaves']], rotation=90)\n",
    "axs[1].set_yticks(range(len(dn['leaves'])))\n",
    "axs[1].set_yticklabels([genotypes[i] if i < len(genotypes) else '1001' for i in dn['leaves']])\n",
    "fig.savefig('figs/mutation_spectra_heatmap.png', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# avg_bqs = df_tmp.bq.apply(lambda x: sum([ord(b) - 33 for b in x]) / len(x))\n",
    "# fig, ax = plt.subplots(figsize=(6, 3))\n",
    "# ax.hist(avg_bqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fig = plt.figure(constrained_layout=True, figsize=(6, 6))\n",
    "# gs = fig.add_gridspec(2, 2, width_ratios=(4, 1), height_ratios=(1, 4))\n",
    "# ax_s = fig.add_subplot(gs[1,0])\n",
    "# ax_s.scatter(avg_bqs, var_pos, alpha=0.1)\n",
    "# fig.add_subplot(gs[0,0], sharex=ax_s).hist(avg_bqs, bins=range(42))\n",
    "# fig.add_subplot(gs[1,1], sharey=ax_s).hist(var_pos, orientation='horizontal', bins=range(151))\n",
    "# fig.suptitle('Duplex seq base quality and position in read of putative mutations')\n",
    "# ax_s.set_xlabel('Base quality')\n",
    "# ax_s.set_ylabel('Position in read')\n",
    "# fig.savefig('figs/duplex_seq_base_quality_and_position_of_variants.png', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# control_bq_hist = np.zeros(45, dtype='i')\n",
    "# control_mq_hist = np.zeros(45, dtype='i')\n",
    "\n",
    "# aln = pysam.AlignmentFile('../data/align/uvtest_UVC_C1_filtered.bam', 'rb')\n",
    "\n",
    "# for i, read in tqdm(enumerate(aln.fetch())):\n",
    "#     if i > 100000:\n",
    "#         break\n",
    "#     if not read.is_proper_pair:\n",
    "#         continue\n",
    "    \n",
    "#     for bq in read.query_qualities:\n",
    "#         control_bq_hist[bq] += 1\n",
    "    \n",
    "#     if not read.get_tag('MD').isdigit():\n",
    "#         control_mq_hist[read.mapping_quality] += 1\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mut_bq_hist = np.zeros(45, dtype='i')\n",
    "# mut_mq_hist = np.zeros(45, dtype='i')\n",
    "\n",
    "# for r in df_all.itertuples():\n",
    "#     for bq in r.bq:\n",
    "#         mut_bq_hist[ord(bq) - 33] += 1\n",
    "#     for mq in r.mq:\n",
    "#         mut_mq_hist[ord(mq) - 33] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fig, axs = plt.subplots(2, 2, sharey=True, sharex=True)\n",
    "# axs[0][0].bar(range(len(control_bq_hist)), control_bq_hist / sum(control_bq_hist))\n",
    "# axs[1][0].bar(range(len(mut_bq_hist)), mut_bq_hist / sum(mut_bq_hist))\n",
    "# axs[0][1].bar(range(len(control_mq_hist)), control_mq_hist / sum(control_mq_hist))\n",
    "# axs[1][1].bar(range(len(mut_mq_hist)), mut_mq_hist / sum(mut_mq_hist))\n",
    "# axs[0][0].set_title('BQ')\n",
    "# axs[0][0].set_ylabel('Null distribution')\n",
    "# axs[0][1].set_title('MQ')\n",
    "# axs[1][0].set_ylabel('Distribution at mutations')\n",
    "# fig.suptitle('BQ and MQ distributions')\n",
    "# fig.savefig('figs/mq_and_bq_of_mutated_reads.png', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_cols = 'chrom source kind start end score strand phase att'.split()\n",
    "df_gff = pd.read_table('data/ref/ref.gff', names=gff_cols) # can get the gff from https://www.arabidopsis.org/api/download-files/download?filePath=Genes/TAIR10_genome_release/TAIR10_gff3/TAIR10_GFF3_genes_transposons.gff\n",
    "df_gff.start -= 1 # convert gff to 0-base half open"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = gtools.load_genome(genome_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rep_gene_models = set()\n",
    "with open('data/ref/TAIR10_representative_gene_models.txt', 'r') as f: # can get this from https://www.arabidopsis.org/api/download-files/download?filePath=Genes/TAIR10_genome_release/TAIR10_gene_lists/TAIR10_representative_gene_models\n",
    "    for l in f:\n",
    "        if l[:2] != 'AT':\n",
    "            continue\n",
    "        rep_gene_models.add(l.strip())\n",
    "print(len(rep_gene_models))\n",
    "\n",
    "# there are also representative gene models for transposable elements and pseudogenes, so restrict the representative models to only protein coding genes\n",
    "protein_genes = set([x.split(\"ID=\")[1].split(\";\")[0] for x in df_gff[df_gff.kind == \"gene\"].att])\n",
    "rep_gene_models = {x for x in rep_gene_models if x.split('.')[0] in protein_genes}\n",
    "assert len(protein_genes) == len(rep_gene_models)\n",
    "len(protein_genes), len(rep_gene_models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reverse complement a sequence\n",
    "comp = {'A':'T', 'T':'A', 'C':'G', 'G':'C'}\n",
    "def rc(seq):\n",
    "    ret = ''\n",
    "    for nuc in reversed(seq):\n",
    "        ret += comp[nuc]\n",
    "    return ret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# identify all synonymous and nonsynonymous sites in the genome\n",
    "s_sites = {chrom:np.zeros(chrom_sizes[chrom], dtype=bool) for chrom in chrom_sizes}\n",
    "ns_sites = {chrom:np.zeros(chrom_sizes[chrom], dtype=bool) for chrom in chrom_sizes}\n",
    "\n",
    "df_cds = df_gff[(df_gff.kind == 'CDS')]\n",
    "for idx, row in tqdm(enumerate(df_cds.itertuples()), total=len(df_cds)):\n",
    "    if row.chrom not in chrom_sizes:\n",
    "        continue\n",
    "    \n",
    "    # only consider the representative gene model (transcript)\n",
    "    trans = row.att.split('Parent=')[1].split(',')[0]\n",
    "    if trans not in rep_gene_models:\n",
    "        continue\n",
    "    \n",
    "    \n",
    "    if row.strand == '+':\n",
    "        for i, pos in enumerate(range(row.start, row.end)):\n",
    "            syn = False\n",
    "            codon_phase = (i - int(row.phase)) % 3\n",
    "            if codon_phase == 2: # if the third base of a codon\n",
    "                if i == 0: # if the first base of the cds element (exon)\n",
    "                    prev_end = df_cds.iloc[idx - 1].end\n",
    "                    codon = genome[row.chrom][prev_end - 2:prev_end]\n",
    "                elif i == 1: # if the second base of the cds element (exon)\n",
    "                    prev_end = df_cds.iloc[idx - 1].end\n",
    "                    codon = genome[row.chrom][prev_end - 1] + genome[row.chrom][pos - 1]\n",
    "                else:\n",
    "                    codon = genome[row.chrom][pos - 2:pos]\n",
    "                \n",
    "                if codon in {'CT', 'GT', 'TC', 'CC', 'AC', 'GC', 'CG', 'GG'}:\n",
    "                    syn = True\n",
    "            \n",
    "            if syn:\n",
    "                s_sites[row.chrom][pos] = 1\n",
    "            else:\n",
    "                ns_sites[row.chrom][pos] = 1\n",
    "    else: # do the same thing for the reverse strand\n",
    "        for i, pos in enumerate(range(row.end - 1, row.start - 1, -1)):\n",
    "            syn = False\n",
    "            codon_phase = (i - int(row.phase)) % 3\n",
    "            if codon_phase == 2:\n",
    "                if i == 0:\n",
    "                    next_start = df_cds.iloc[idx + 1].start\n",
    "                    codon = rc(genome[row.chrom][next_start:next_start + 2])\n",
    "                elif i == 1:\n",
    "                    next_start = df_cds.iloc[idx + 1].start\n",
    "                    codon = rc(genome[row.chrom][pos + 1] + genome[row.chrom][next_start])\n",
    "                else:\n",
    "                    codon = rc(genome[row.chrom][pos + 1:pos + 3])\n",
    "                \n",
    "                if codon in {'CT', 'GT', 'TC', 'CC', 'AC', 'GC', 'CG', 'GG'}:\n",
    "                    syn = True\n",
    "            \n",
    "            if syn:\n",
    "                s_sites[row.chrom][pos] = 1\n",
    "            else:\n",
    "                ns_sites[row.chrom][pos] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # make bed files to visually confirm sites are getting the correct label\n",
    "# ns_f = open('tmp/ns_sites.bed', 'w')\n",
    "# s_f = open('tmp/s_sites.bed', 'w')\n",
    "# for i in range(100000):\n",
    "#     if ns_sites['Chr1'][i] == 1:\n",
    "#         ns_f.write(f'Chr1\\t{str(i)}\\t{str(i + 1)}\\n')\n",
    "#     if s_sites['Chr1'][i] == 1:\n",
    "#         s_f.write(f'Chr1\\t{str(i)}\\t{str(i + 1)}\\n')\n",
    "# ns_f.close()\n",
    "# s_f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the mutation count and coverage of synonymous and nonsynonymous sites for each sample\n",
    "dnds = {s: {'s_count':0, 'ns_count':0, 'other_count':0, 's_cov':0, 'ns_cov':0} for s in sample_names}\n",
    "for s in tqdm(sample_names):\n",
    "    # load the callable coverage for the sample\n",
    "    call_cov = dict()\n",
    "    for chrom in chrom_sizes:\n",
    "        if 'MA lines' or '1001' in s:\n",
    "            call_cov[chrom] = np.load(f'{cov_prefixes[\"untreated_1\"]}{chrom}.npy') > 0\n",
    "        elif 'Ler-0' in s:\n",
    "            call_cov[chrom] = np.load(f'{cov_prefixes[sample]}{chrom}.npy') > 0\n",
    "        else:\n",
    "            call_cov[chrom] = np.load(f'{cov_prefixes[sample]}{chrom}.npy')\n",
    "    \n",
    "    \n",
    "    # count the mutations at synonymous and nonsynonymous sites\n",
    "    s_count = 0\n",
    "    ns_count = 0\n",
    "    other_count = 0\n",
    "    for row in dfs_somatic[s].itertuples():\n",
    "        if s_sites[row.chrom][row.pos]:\n",
    "            s_count += 1\n",
    "        elif ns_sites[row.chrom][row.pos]:\n",
    "            ns_count += 1\n",
    "        else:\n",
    "            other_count += 1\n",
    "    \n",
    "    # count the coverage at synonymous and nonsynonymous sites\n",
    "    s_cov = 0\n",
    "    ns_cov = 0\n",
    "    for chrom in chrom_sizes:\n",
    "        s_cov += np.sum(s_sites[row.chrom] * call_cov[row.chrom])\n",
    "        ns_cov += np.sum(ns_sites[row.chrom] * call_cov[row.chrom])\n",
    "    \n",
    "    # add values to dnds\n",
    "    dnds[s]['s_count'] = s_count\n",
    "    dnds[s]['ns_count'] = ns_count\n",
    "    dnds[s]['other_count'] = other_count\n",
    "    dnds[s]['s_cov'] = s_cov\n",
    "    dnds[s]['ns_cov'] = ns_cov\n",
    "    \n",
    "    # add values to 'all somatic'\n",
    "    # if 'Ler-0-germline' not in s and '1001' not in s and 'MA lines' not in s:\n",
    "    #     dnds['all somatic']['s_count'] += s_count\n",
    "    #     dnds['all somatic']['ns_count'] += ns_count\n",
    "    #     dnds['all somatic']['s_cov'] += s_cov\n",
    "    #     dnds['all somatic']['ns_cov'] += ns_cov\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add a rate ratio to dnds\n",
    "cannot_compute = []\n",
    "for s in dnds:\n",
    "    if dnds[s]['s_count'] == 0:\n",
    "        cannot_compute.append(s)\n",
    "        dnds[s]['ratio'] = np.nan\n",
    "        continue\n",
    "    dnds[s]['ratio'] = (dnds[s]['ns_count'] / dnds[s]['ns_cov']) / (dnds[s]['s_count'] / dnds[s]['s_cov'])\n",
    "print(f'could not compute dnds for {cannot_compute}, no synonymous mutations')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lower = dict()\n",
    "upper = dict()\n",
    "for g in dnds:\n",
    "    boots = []\n",
    "    reps = 10000 if dnds[g]['ns_count'] < 10000 else 20 # number of bootstrap replicates, use a small number for the genotypes with a ton of mutations\n",
    "    for i in tqdm(range(reps)):\n",
    "        # randomly sample the same number of mutations with replacement\n",
    "        total_count = dnds[g]['s_count'] + dnds[g]['ns_count'] + dnds[g]['other_count']\n",
    "        choices = random.choices([0, 1, 2], weights=[dnds[g]['s_count'] / total_count, dnds[g]['ns_count'] / total_count, dnds[g]['other_count'] / total_count], k=total_count)\n",
    "        \n",
    "        # same as above, but don't consider the mutations in non-CDS regions; I tried both ways and didn't see any difference\n",
    "        # total_count = dnds[g]['s_count'] + dnds[g]['ns_count']\n",
    "        # choices = random.choices([0, 1], weights=[dnds[g]['s_count'] / total_count, dnds[g]['ns_count'] / total_count], k=total_count)\n",
    "        \n",
    "        # calculate the dnds of this sample\n",
    "        boot_s = choices.count(0)\n",
    "        boot_ns = choices.count(1)\n",
    "        if boot_s > 0:\n",
    "            boots.append((boot_ns / dnds[g]['ns_cov']) / (boot_s / dnds[g]['s_cov']))\n",
    "        else:\n",
    "            boots.append(2)\n",
    "    \n",
    "    # pick the 5th and 95th percentile dnds as the lower and upper confidence interval\n",
    "    boots.sort()\n",
    "    lower[g] = boots[int(len(boots) * 0.05)]\n",
    "    upper[g] = boots[int(len(boots) * 0.95)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_somatic = ['untreated', 'UVB', 'UVC']\n",
    "vis_genotypes = ['MA lines', 'Ler-0', '1001']\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(4, 6))\n",
    "ax.hlines(1, -1, len(dnds) + 1, color='black')\n",
    "\n",
    "sems = dict()\n",
    "a = []\n",
    "for g in all_somatic + vis_genotypes:\n",
    "    ratios = [dnds[s]['ratio'] for s in g_to_s(g)]\n",
    "    if np.nan not in ratios:\n",
    "        sems[g] = np.std(ratios) / (len(ratios) ** (1/2))\n",
    "    else:\n",
    "        print(f'WARNING: undefined dnds values for some samples in {g}, bootstrapping the sem')\n",
    "        boots = []\n",
    "        for b in range(2000):\n",
    "            chosen = random.choices(g_to_s(g), k=len(g_to_s(g)))\n",
    "            ns_count = sum([dnds[s]['ns_count'] for s in chosen])\n",
    "            ns_cov = sum([dnds[s]['ns_cov'] for s in chosen])\n",
    "            s_count = sum([dnds[s]['s_count'] for s in chosen])\n",
    "            s_cov = sum([dnds[s]['s_cov'] for s in chosen])\n",
    "            ratio = (ns_count / ns_cov) / (s_count / s_cov)\n",
    "            boots.append(ratio)\n",
    "        sems[g] = np.std(boots)\n",
    "        \n",
    "    if g in all_somatic:\n",
    "        a += ratios\n",
    "sems['all somatic'] = np.std(a) / (len(a) ** (1/2))\n",
    "\n",
    "ns_count, ns_cov, s_count, s_cov = 0, 0, 0, 0\n",
    "for g in all_somatic:\n",
    "    ns_count += sum([dnds[s]['ns_count'] for s in g_to_s(g)])\n",
    "    ns_cov += sum([dnds[s]['ns_cov'] for s in g_to_s(g)])\n",
    "    s_count += sum([dnds[s]['s_count'] for s in g_to_s(g)])\n",
    "    s_cov += sum([dnds[s]['s_cov'] for s in g_to_s(g)])\n",
    "ratio = (ns_count / ns_cov) / (s_count / s_cov)\n",
    "ax.bar(0, ratio, yerr=sems['all somatic'], color='tab:grey')\n",
    "\n",
    "for g in all_somatic:\n",
    "    for s in g_to_s(g):\n",
    "        ratio = (dnds[s]['ns_count'] / dnds[s]['ns_cov']) / (dnds[s]['s_count'] / dnds[s]['s_cov'])\n",
    "        ax.scatter(-0.25 + random.random() * 0.5, ratio, color=gen_cmap(g))\n",
    "\n",
    "for i, g in enumerate(vis_genotypes):\n",
    "    print(g)\n",
    "    ns_count = sum([dnds[s]['ns_count'] for s in g_to_s(g)])\n",
    "    ns_cov = sum([dnds[s]['ns_cov'] for s in g_to_s(g)])\n",
    "    s_count = sum([dnds[s]['s_count'] for s in g_to_s(g)])\n",
    "    s_cov = sum([dnds[s]['s_cov'] for s in g_to_s(g)])\n",
    "    ratio = (ns_count / ns_cov) / (s_count / s_cov)\n",
    "    print(g, ns_count, ns_cov, s_count, s_cov)\n",
    "    ax.bar(i + 1, ratio, yerr=sems[g], color=gen_cmap(g))\n",
    "\n",
    "ax.set_xlim(-0.5, len(vis_genotypes) + 0.5)\n",
    "# ax.set_ylim(0, 2)\n",
    "ax.set_xticks(range(len(vis_genotypes) + 1))\n",
    "ax.set_xticklabels(['all somatic'] + vis_genotypes, rotation=45, ha='right')\n",
    "ax.set_title('dN/dS ratios by sample')\n",
    "ax.set_ylabel('dN/dS')\n",
    "fig.savefig('figs/dnds_across_samples.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Look for stranded results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f_gene = {chrom:np.zeros(chrom_sizes[chrom], dtype=bool) for chrom in chrom_sizes}\n",
    "r_gene = {chrom:np.zeros(chrom_sizes[chrom], dtype=bool) for chrom in chrom_sizes}\n",
    "\n",
    "for row in df_gff[df_gff.kind == 'mRNA'].itertuples():\n",
    "    if row.att.split('ID=')[1].split(';')[0] not in rep_gene_models or row.chrom in ['ChrC', 'ChrM']:\n",
    "        continue\n",
    "    \n",
    "    if row.strand == '+':\n",
    "        f_gene[row.chrom][row.start:row.end] = True\n",
    "    else:\n",
    "        r_gene[row.chrom][row.start:row.end] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "simple_contexts = ['CA', 'CG', 'CT', 'TA', 'TC', 'TG', 'del', 'ins']\n",
    "pyr_forward_s = {s:{c:0 for c in simple_contexts} for s in sample_names}\n",
    "pyr_reverse_s = {s:{c:0 for c in simple_contexts} for s in sample_names}\n",
    "pyr_forward_g = {g:{c:0 for c in simple_contexts} for g in genotypes}\n",
    "pyr_reverse_g = {g:{c:0 for c in simple_contexts} for g in genotypes}\n",
    "\n",
    "for s in tqdm(sample_names):\n",
    "    for row in dfs_somatic[s].itertuples():\n",
    "        if f_gene[row.chrom][row.pos] and not r_gene[row.chrom][row.pos]:\n",
    "            reverse = False\n",
    "        elif r_gene[row.chrom][row.pos] and not f_gene[row.chrom][row.pos]:\n",
    "            reverse = True\n",
    "        else:\n",
    "            # if s == 'Col-0-8':\n",
    "            #     print(f'{row.chrom}:{row.pos} {row.ref}>{row.alt} not in gene')\n",
    "            continue\n",
    "        \n",
    "        if row.ref == '*':\n",
    "            context = 'ins'\n",
    "        elif row.alt == '*':\n",
    "            context = 'del'\n",
    "        elif row.ref == 'C' or row.ref == 'T':\n",
    "            context = row.ref + row.alt\n",
    "        elif row.ref == 'A' or row.ref == 'G':\n",
    "            context = comp[row.ref] + comp[row.alt]\n",
    "            reverse = not reverse\n",
    "        \n",
    "        g = s_to_g(s)\n",
    "        if reverse:\n",
    "            pyr_reverse_s[s][context] += 1\n",
    "            pyr_reverse_g[g][context] += 1\n",
    "        else:\n",
    "            pyr_forward_s[s][context] += 1\n",
    "            pyr_forward_g[g][context] += 1\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vis_genotypes = genotypes\n",
    "vis_samples = [s for s in sample_names if s_to_g(s) in ['Col-0', 'UVB', 'UVC']]\n",
    "\n",
    "sems = dict()\n",
    "for g in genotypes:\n",
    "    try:\n",
    "        sems[g] = np.std([pyr_forward_s[s]['CT'] / pyr_reverse_s[s]['CT'] for s in g_to_s(g)]) / (len(g_to_s(g)) ** (1/2))\n",
    "    except ZeroDivisionError:\n",
    "        print(f'WARNING: some samples in {g} have undefined ratios, calculating a bootstrap sem')\n",
    "        try:\n",
    "            boot_means = []\n",
    "            for b in range(2000):\n",
    "                f = sum(random.choices([pyr_forward_s[s]['CT'] for s in g_to_s(g)], k=len(g_to_s(g))))\n",
    "                r = sum(random.choices([pyr_reverse_s[s]['CT'] for s in g_to_s(g)], k=len(g_to_s(g))))\n",
    "                boot_means.append(f / r)\n",
    "            sems[g] = np.std(boot_means)\n",
    "        except ZeroDivisionError:\n",
    "            print(f'WARNING: some bootstrpas in {g} also have undefinded ratios, setting sem to 0')\n",
    "            sems[g] = 0\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3.5, 4))\n",
    "for i, g in enumerate(vis_genotypes):\n",
    "    forward = pyr_forward_g[g]['CT']\n",
    "    reverse = pyr_reverse_g[g]['CT']\n",
    "    forward_cinf = stats.binom.interval(0.95, forward + reverse, forward / (forward + reverse))\n",
    "    ratio = forward / reverse\n",
    "    low_err = ratio - (forward_cinf[0] / (forward + reverse - forward_cinf[0]))\n",
    "    high_err = (forward_cinf[1] / (forward + reverse - forward_cinf[1])) - ratio\n",
    "    ax.bar(i, ratio, yerr=sems[g], color=gen_cmap(g), alpha=0.8)\n",
    "for s in vis_samples:\n",
    "    ratio = pyr_forward_s[s]['CT'] / pyr_reverse_s[s]['CT']\n",
    "    ax.scatter(vis_genotypes.index(s_to_g(s)) + random.random() * 0.5 - 0.25, ratio, color=gen_cmap(s_to_g(s)))\n",
    "ax.hlines(1, -10, 10, color='black')\n",
    "ax.set_xlim(-0.75, len(vis_genotypes) - 0.25)\n",
    "ax.set_xticks(range(len(vis_genotypes)), vis_genotypes)\n",
    "ax.set_ylabel('nontemplate / template C>T mutations')\n",
    "fig.savefig('figs/strand_bias_of_ct_mutations.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for g in vis_genotypes:\n",
    "    try:\n",
    "        print(g, stats.ttest_1samp([pyr_forward_s[s]['CT'] / pyr_reverse_s[s]['CT'] for s in g_to_s(g)], 1, alternative='greater').pvalue * 4)\n",
    "    except ZeroDivisionError:\n",
    "        print(g, 'error')\n",
    "        continue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Mutation rate by genomic region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "regions = { # these files are generated by bed_regions.ipynb\n",
    "    'exon':'data/region/exon.bed',\n",
    "    'intron':'data/region/intron.bed',\n",
    "    'te':'data/region/te.bed',\n",
    "    'promoter':'data/region/promoter.bed',\n",
    "    'intergenic':'data/region/intergenic.bed',\n",
    "    'acr':'data/region/acrs.bed',\n",
    "    'gene':'data/region/protein_mRNA.bed',\n",
    "    'pericentromere':'data/region/pericentromere.bed',\n",
    "    'met cg':'data/region/met_cg.bed',\n",
    "    'met chg':'data/region/met_chg.bed',\n",
    "    'met chh':'data/region/met_chh.bed',\n",
    "    'unmet cg':'data/region/unmet_cg.bed',\n",
    "    'unmet chg':'data/region/unmet_chg.bed',\n",
    "    'unmet chh':'data/region/unmet_chh.bed',\n",
    "    # 'near met':'data/region/near_met.bed',\n",
    "}\n",
    "\n",
    "region_arrays = {r:{chrom:np.zeros(chrom_sizes[chrom], dtype=bool) for chrom in chrom_sizes} for r in regions}\n",
    "for r in tqdm(region_arrays):\n",
    "    df_bed = pd.read_table(regions[r], names='chrom start end'.split(), usecols=list(range(3)))\n",
    "    df_bed = df_bed[df_bed.chrom.apply(lambda c: c in chrom_sizes)]\n",
    "    for row in df_bed.itertuples():\n",
    "        region_arrays[r][row.chrom][row.start:row.end] = True\n",
    "    del df_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add extra region arrays using the existing ones\n",
    "\n",
    "region_arrays['pericentromeric genes'] = {chrom:region_arrays['gene'][chrom] * region_arrays['pericentromere'][chrom] for chrom in chrom_sizes}\n",
    "region_arrays['distal genes'] = {chrom:region_arrays['gene'][chrom] * ~region_arrays['pericentromere'][chrom] for chrom in chrom_sizes}\n",
    "region_arrays['pericentromeric tes'] = {chrom:region_arrays['te'][chrom] * region_arrays['pericentromere'][chrom] for chrom in chrom_sizes}\n",
    "region_arrays['distal tes'] = {chrom:region_arrays['te'][chrom] * ~region_arrays['pericentromere'][chrom] for chrom in chrom_sizes}\n",
    "\n",
    "region_arrays['genic met cg'] = {chrom:region_arrays['met cg'][chrom] * region_arrays['gene'][chrom] for chrom in chrom_sizes}\n",
    "region_arrays['genic unmet cg'] = {chrom:region_arrays['unmet cg'][chrom] * region_arrays['gene'][chrom] for chrom in chrom_sizes}\n",
    "region_arrays['nongenic met cg'] = {chrom:region_arrays['met cg'][chrom] * ~region_arrays['gene'][chrom] for chrom in chrom_sizes}\n",
    "region_arrays['nongenic unmet cg'] = {chrom:region_arrays['unmet cg'][chrom] * ~region_arrays['gene'][chrom] for chrom in chrom_sizes}\n",
    "region_arrays['nongenic met chg'] = {chrom:region_arrays['met chg'][chrom] * ~region_arrays['gene'][chrom] for chrom in chrom_sizes}\n",
    "region_arrays['nongenic unmet chg'] = {chrom:region_arrays['unmet chg'][chrom] * ~region_arrays['gene'][chrom] for chrom in chrom_sizes}\n",
    "region_arrays['nongenic met chh'] = {chrom:region_arrays['met chh'][chrom] * ~region_arrays['gene'][chrom] for chrom in chrom_sizes}\n",
    "region_arrays['nongenic unmet chh'] = {chrom:region_arrays['unmet chh'][chrom] * ~region_arrays['gene'][chrom] for chrom in chrom_sizes}\n",
    "\n",
    "region_arrays['genic acr'] = {chrom:region_arrays['acr'][chrom] * region_arrays['gene'][chrom] for chrom in chrom_sizes}\n",
    "region_arrays['genic nonacr'] = {chrom:~region_arrays['acr'][chrom] * region_arrays['gene'][chrom] for chrom in chrom_sizes}\n",
    "region_arrays['nongenic acr'] = {chrom:region_arrays['acr'][chrom] * ~region_arrays['gene'][chrom] for chrom in chrom_sizes}\n",
    "region_arrays['nongenic nonacr'] = {chrom:~region_arrays['acr'][chrom] * ~region_arrays['gene'][chrom] for chrom in chrom_sizes}\n",
    "\n",
    "# region_arrays['exon nomet'] = {chrom:region_arrays['exon'][chrom] * (region_arrays['at'][chrom] + region_arrays['unmet cg'][chrom] + region_arrays['unmet chg'][chrom] + region_arrays['unmet chh'][chrom]) for chrom in chrom_sizes}\n",
    "# region_arrays['intron nomet'] = {chrom:region_arrays['intron'][chrom] * (region_arrays['at'][chrom] + region_arrays['unmet cg'][chrom] + region_arrays['unmet chg'][chrom] + region_arrays['unmet chh'][chrom]) for chrom in chrom_sizes}\n",
    "# region_arrays['te nomet'] = {chrom:region_arrays['te'][chrom] * (region_arrays['at'][chrom] + region_arrays['unmet cg'][chrom] + region_arrays['unmet chg'][chrom] + region_arrays['unmet chh'][chrom]) for chrom in chrom_sizes}\n",
    "# region_arrays['promoter nomet'] = {chrom:region_arrays['promoter'][chrom] * (region_arrays['at'][chrom] + region_arrays['unmet cg'][chrom] + region_arrays['unmet chg'][chrom] + region_arrays['unmet chh'][chrom]) for chrom in chrom_sizes}\n",
    "# region_arrays['intergenic nomet'] = {chrom:region_arrays['intergenic'][chrom] * (region_arrays['at'][chrom] + region_arrays['unmet cg'][chrom] + region_arrays['unmet chg'][chrom] + region_arrays['unmet chh'][chrom]) for chrom in chrom_sizes}\n",
    "\n",
    "# region_arrays['exon yc'] = {chrom:region_arrays['exon'][chrom] * yc_mask[chrom] for chrom in chrom_sizes}\n",
    "# region_arrays['intron yc'] = {chrom:region_arrays['intron'][chrom] * yc_mask[chrom] for chrom in chrom_sizes}\n",
    "# region_arrays['te yc'] = {chrom:region_arrays['te'][chrom] * yc_mask[chrom] for chrom in chrom_sizes}\n",
    "# region_arrays['promoter yc'] = {chrom:region_arrays['promoter'][chrom] * yc_mask[chrom] for chrom in chrom_sizes}\n",
    "# region_arrays['intergenic yc'] = {chrom:region_arrays['intergenic'][chrom] * yc_mask[chrom] for chrom in chrom_sizes}\n",
    "\n",
    "\n",
    "# region_arrays['cg'] = {chrom:np.array([n == 'C' or n == 'G' for n in genome[chrom]]) for chrom in chrom_sizes}\n",
    "# region_arrays['at'] = {chrom:np.array([n == 'A' or n == 'T' for n in genome[chrom]]) for chrom in chrom_sizes}\n",
    "\n",
    "# for region in 'exon intron te promoter intergenic'.split():\n",
    "    # region_arrays[f'{region} cg'] = {chrom:region_arrays[region][chrom] * region_arrays['cg'][chrom] for chrom in chrom_sizes}\n",
    "    # region_arrays[f'{region} at'] = {chrom:region_arrays[region][chrom] * region_arrays['at'][chrom] for chrom in chrom_sizes}\n",
    "    # region_arrays[f'{region} pericentromere'] = {chrom:region_arrays[region][chrom] * region_arrays['pericentromere'][chrom] for chrom in chrom_sizes}\n",
    "    # region_arrays[f'{region} distal'] = {chrom:region_arrays[region][chrom] * ~region_arrays['pericentromere'][chrom] for chrom in chrom_sizes}\n",
    "    # region_arrays[f'{region} met'] = {chrom:region_arrays[region][chrom] * region_arrays['met c'][chrom] for chrom in chrom_sizes}\n",
    "    # region_arrays[f'{region} unmet'] = {chrom:region_arrays[region][chrom] * region_arrays['unmet c'][chrom] for chrom in chrom_sizes}\n",
    "    # region_arrays[f'{region} acr'] = {chrom:region_arrays[region][chrom] * region_arrays['acr'][chrom] for chrom in chrom_sizes}\n",
    "    # region_arrays[f'{region} nonacr'] = {chrom:region_arrays[region][chrom] * ~region_arrays['acr'][chrom] for chrom in chrom_sizes}\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "del region_arrays['met cg'], region_arrays['met chg'], region_arrays['met chh'], region_arrays['unmet cg'], region_arrays['unmet chg'], region_arrays['unmet chh']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def array_to_bed(array, bed):\n",
    "    fout = open(bed, 'w')\n",
    "    for chrom in chrom_sizes:\n",
    "        region_start = None\n",
    "        for i, value in enumerate(array[chrom]):\n",
    "            if value and region_start is None:\n",
    "                region_start = i\n",
    "            elif value:\n",
    "                pass\n",
    "            elif region_start is None:\n",
    "                pass\n",
    "            else:\n",
    "                fout.write(f'{chrom}\\t{region_start}\\t{i}\\n')\n",
    "                region_start = None\n",
    "    fout.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# array_to_bed(region_arrays['intron acr'], 'tmp/region_intron_acr.bed')\n",
    "# array_to_bed(region_arrays['intron nonacr'], 'tmp/region_intron_nonacr.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# count the number of mutations in each element type for each sample\n",
    "elem_counts = { # indexed as elem_counts[thing][sample/genotype index][element type] = mutations in element\n",
    "    'ind':{s:{r:0 for r in region_arrays} for s in dfs_somatic}, # individual samples\n",
    "    'gen':{g:{r:0 for r in region_arrays} for g in genotypes}, # sum over all samples with the same genotype\n",
    "    'low':{g:{r:0 for r in region_arrays} for g in genotypes}, # lower confidence interval by genotype\n",
    "    'high':{g:{r:0 for r in region_arrays} for g in genotypes}, # upper confidence interval by genotype\n",
    "    'gen_uniform':{g:0 for g in genotypes}\n",
    "}\n",
    "for sample, df in tqdm(dfs_somatic.items()):\n",
    "    genotype = s_to_g(sample)\n",
    "    for row in df.itertuples():\n",
    "        for r in region_arrays:\n",
    "            if region_arrays[r][row.chrom][row.pos]:\n",
    "                c = 1\n",
    "                # if row.ref == 'C' or row.ref == 'G': # adjust for different rates in CG vs AT sites\n",
    "                #     c = 1 / cg_at_ratio[genotype]\n",
    "                elem_counts['ind'][sample][r] += c\n",
    "                elem_counts['gen'][genotype][r] += c\n",
    "    # elem_counts['gen_uniform'][genotype] += sum((df.ref != 'C') & (df.ref != 'G')) + sum((df.ref == 'C') | (df.ref == 'G')) / cg_at_ratio[genotype]\n",
    "    elem_counts['gen_uniform'][genotype] += len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate 95% confidence intervals using binomial distributions\n",
    "for g in tqdm(genotypes):\n",
    "    for t in region_arrays:\n",
    "        muts_in_elem = elem_counts['gen'][g][t]\n",
    "        total_muts = elem_counts['gen_uniform'][g]\n",
    "        interval = stats.binom.interval(0.95, total_muts, muts_in_elem / total_muts)\n",
    "    \n",
    "        elem_counts['low'][g][t] = interval[0]\n",
    "        elem_counts['high'][g][t] = interval[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate chi-squared p-values that each genotype differs from Col-0. \n",
    "# note: these don't really have much meaning\n",
    "p_vals = []\n",
    "for g in genotypes:\n",
    "    assert elem_counts['gen'][g].keys() == region_arrays.keys()\n",
    "    act = list(elem_counts['gen'][g].values())\n",
    "    exp = list(elem_counts['gen'][genotypes[0]].values())\n",
    "    xs = stats.chisquare(act, [x * (sum(act) / sum(exp)) for x in exp])\n",
    "    p_vals.append(xs.pvalue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "elem_cov = {\n",
    "    'ind':{s:{t:0 for t in region_arrays} for s in sample_names}, # individual samples\n",
    "    'gen':{g:{t:0 for t in region_arrays} for g in genotypes}, # sum over all samples with the same genotype\n",
    "    'uniform':{g:0 for g in genotypes}\n",
    "}\n",
    "\n",
    "wgs_cov_arrs = {chrom:np.load(f'{cov_prefixes[\"untreated_1\"]}{chrom}.npy') > 0 for chrom in chrom_sizes}\n",
    "wgs_cov_sum = sum([np.sum(wgs_cov_arrs[chrom]) for chrom in chrom_sizes])\n",
    "# first_ma_sample = [x for x in sample_names if 'MA lines' in x or '1001' in x][0] if 'MA lines' in genotypes else None\n",
    "for s in tqdm(sample_names):\n",
    "    g = s_to_g(s)\n",
    "    \n",
    "    # if 'MA lines' in s and s != first_ma_sample: # skip recalculating MA line coverage since they're all the same FIXME this code changed the results once, but it was probably because non-zero conflict rates were entered for the MA lines\n",
    "    #     for t in region_arrays:\n",
    "    #         first_cov = elem_cov['ind'][first_ma_sample][t]\n",
    "    #         elem_cov['ind'][s][t] = first_cov\n",
    "    #         elem_cov['gen'][g][t] += first_cov\n",
    "    #     elem_cov['uniform'][g] += wgs_cov_sum\n",
    "    #     continue\n",
    "    \n",
    "    for chrom in chrom_sizes:\n",
    "        if 'Ler-0' in s:\n",
    "            cov_arr = np.load(f'{cov_prefixes[s]}{chrom}.npy')\n",
    "            cov_arr = cov_arr > 0\n",
    "        elif 'MA lines' in s or '1001' in s:\n",
    "            cov_arr = wgs_cov_arrs[chrom]\n",
    "        else:\n",
    "            cov_arr = np.load(f'{cov_prefixes[s]}{chrom}.npy')\n",
    "        for t in region_arrays:\n",
    "            c = np.sum(region_arrays[t][chrom] * cov_arr) * (1 - conflict_rates[sample_names.index(s)])\n",
    "            elem_cov['ind'][s][t] += c\n",
    "            elem_cov['gen'][g][t] += c\n",
    "        elem_cov['uniform'][g] += np.sum(cov_arr) * (1 - conflict_rates[sample_names.index(s)])\n",
    "del wgs_cov_arrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert the proportion of mutations in each element to a rate using the callable coverage\n",
    "elem_rates = { # indexed as elem_counts[thing][sample/genotype index][element type] = mutation rate in element (mutations / callable bp in first sample)\n",
    "    'ind':{s:{t:0 for t in region_arrays} for s in sample_names}, # individual samples\n",
    "    'gen':{g:{t:0 for t in region_arrays} for g in genotypes}, # sum over all samples with the same genotype\n",
    "    'low':{g:{t:0 for t in region_arrays} for g in genotypes}, # lower confidence interval by genotype\n",
    "    'high':{g:{t:0 for t in region_arrays} for g in genotypes}, # upper confidence interval by genotype\n",
    "    'gen_uniform': {g:0 for g in genotypes}\n",
    "}\n",
    "\n",
    "for thing in elem_counts:\n",
    "    for sg in elem_counts[thing]: # sg is either the sample or the genotype\n",
    "        if thing != 'gen_uniform':\n",
    "            cov_thing = 'ind' if thing == 'ind' else 'gen'\n",
    "            for t in region_arrays:\n",
    "                elem_rates[thing][sg][t] = elem_counts[thing][sg][t] / elem_cov[cov_thing][sg][t]\n",
    "        else:\n",
    "            elem_rates[thing][sg] = elem_counts[thing][sg] / elem_cov['uniform'][sg]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# normalize the mutation rates in each element to the uniform rate\n",
    "norm_elem_rates = { # indexed as elem_counts[thing][sample/genotype index][element type] = mutation rate in element / genome-wide rate\n",
    "    'ind':{s:{t:0 for t in region_arrays} for s in sample_names}, # individual samples\n",
    "    'gen':{g:{t:0 for t in region_arrays} for g in genotypes}, # sum over all samples with the same genotype\n",
    "    'low':{g:{t:0 for t in region_arrays} for g in genotypes}, # lower confidence interval by genotype\n",
    "    'high':{g:{t:0 for t in region_arrays} for g in genotypes}, # upper confidence interval by genotype\n",
    "    'gen_uniform': {g:0 for g in genotypes}\n",
    "}\n",
    "\n",
    "for thing in elem_counts:\n",
    "    for sg in elem_rates[thing]:\n",
    "        if thing == 'ind': # if looking at an individual sample, normalize by the genotype's uniform rate\n",
    "            g = s_to_g(sg)\n",
    "            # print(f'{sg} h3k9 rate = {elem_rates[thing][sg][t]}, uniform rate = {elem_rates[\"gen_uniform\"][g]}')\n",
    "            norm_elem_rates[thing][sg] = {t:elem_rates[thing][sg][t] / elem_rates['gen_uniform'][g] for t in region_arrays}\n",
    "        elif thing != 'gen_uniform': # normalize the uniform rate by itself, should be 1\n",
    "            norm_elem_rates[thing][sg] = {t:elem_rates[thing][sg][t] / elem_rates['gen_uniform'][sg] for t in region_arrays}\n",
    "        else:\n",
    "            norm_elem_rates[thing][sg] = elem_rates[thing][sg] / elem_rates['gen_uniform'][sg]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vis_genotypes = genotypes\n",
    "unique_samples = [s for s in sample_names if [s_to_g(x) for x in sample_names].count(s_to_g(s)) == 1]\n",
    "vis_samples = [s for s in sample_names if s_to_g(s) in vis_genotypes and s not in unique_samples and s_to_g(s) != 'MA lines']\n",
    "vis_regions = ['exon', 'intron', 'te', 'promoter', 'intergenic']\n",
    "# vis_regions = ['pericentromeric genes', 'distal genes', 'pericentromeric tes', 'distal tes']\n",
    "# vis_regions = ['pericentromeric syn', 'distal syn', 'pericentromeric nonsyn', 'distal nonsyn']\n",
    "# vis_regions = ['genic unmet cg', 'genic met cg', 'nongenic unmet cg', 'nongenic met cg', 'nongenic unmet chg', 'nongenic met chg', 'nongenic unmet chh', 'nongenic met chh']\n",
    "# vis_regions = ['genic acr', 'genic nonacr', 'nongenic acr', 'nongenic nonacr']#, 'nongenic acr unmet distal', 'nongenic nonacr unmet distal']\n",
    "# vis_regions = ['exon acr', 'exon nonacr', 'intron acr', 'intron nonacr', 'te acr', 'te nonacr', 'promoter acr', 'promoter nonacr', 'intergenic acr', 'intergenic nonacr']\n",
    "# vis_regions = ['exon yc', 'intron yc', 'te yc', 'promoter yc', 'intergenic yc']\n",
    "sub_norm_elem_rates = {'ind':dict(), 'gen':dict(), 'low':dict(), 'high':dict()}\n",
    "sample_to_genotype = []\n",
    "gen_colors = []\n",
    "\n",
    "for sample in vis_samples:\n",
    "    sample_genotype = sample.rsplit('-', 1)[0]\n",
    "    sub_norm_elem_rates['ind'][sample] = {region:norm_elem_rates['ind'][sample][region] for region in vis_regions}\n",
    "    \n",
    "for genotype in vis_genotypes:\n",
    "    sub_norm_elem_rates['gen'][genotype] = {region:norm_elem_rates['gen'][genotype][region] for region in vis_regions}\n",
    "    sub_norm_elem_rates['low'][genotype] = {region:norm_elem_rates['low'][genotype][region] for region in vis_regions}\n",
    "    sub_norm_elem_rates['high'][genotype] = {region:norm_elem_rates['high'][genotype][region] for region in vis_regions}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def log2_min(x):\n",
    "    if x == 0:\n",
    "        return -2\n",
    "    else:\n",
    "        return math.log2(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sems = {g:dict() for g in vis_genotypes}\n",
    "for g in vis_genotypes:\n",
    "    for region in vis_regions:\n",
    "        # I tried bootstrapping sems here. It gives smaller values so I decided not to use it. The normal way of calculating sem is\n",
    "        # probably overly conservative though, as the samples have quantized rates which make it impossible to have values right on the mean\n",
    "        if g == 'MA lines': \n",
    "            boot_rates = []\n",
    "            for b in range(2000):\n",
    "                boot_samples = random.choices(g_to_s('MA lines'), k=len(g_to_s('MA lines')))\n",
    "                region_count = sum([elem_counts['ind'][s][region] for s in boot_samples])\n",
    "                total_count = sum([len(dfs_somatic[s]) for s in boot_samples])\n",
    "                region_cov = sum([elem_cov['ind'][s][region] for s in boot_samples])\n",
    "                total_cov = elem_cov['uniform'][g] # the total cov remains the same regardless of which samples were bootstrapped\n",
    "                \n",
    "                relative_rate = (region_count / region_cov) / (total_count / total_cov)\n",
    "                boot_rates.append(relative_rate)\n",
    "            boot_sem = np.std(boot_rates)\n",
    "            normal_sem = np.std([norm_elem_rates[\"ind\"][s][region] for s in g_to_s(g)]) / (len(g_to_s(g)) ** (1/2))\n",
    "            \n",
    "            print(f'{boot_sem:.4f} bootstrapped sem, {normal_sem:.4f} raw sem')\n",
    "            sems[g][region] = normal_sem # just use the normal method instead of the bootstrap estimate\n",
    "        else:\n",
    "            # calculate using replicates\n",
    "            sems[g][region] = np.std([norm_elem_rates['ind'][s][region] for s in g_to_s(g)]) / (len(g_to_s(g)) ** (1/2))\n",
    "\n",
    "            # calculate using binomial (ignore biological differences in mutation distribution between replicates), FIXME I don't think this is right\n",
    "            # count_sem = stats.binom.std(gen_count_sums[g], elem_counts['gen'][g][region] / gen_count_sums[g]) / (gen_count_sums[g] ** (1/2))\n",
    "            # sems[g][region] = (count_sem / elem_cov['gen'][g][region]) / elem_rates['gen_uniform'][g]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot log2 difference of mutation rate relative to uniform (e.g. a value of -1 means there are half as many mutations in the element as expected in a uniform distribution)\n",
    "fig, ax = plt.subplots(figsize=(len(vis_regions) * (len(vis_genotypes) + 1) * 0.2, 3))\n",
    "bars = []\n",
    "for g in vis_genotypes:\n",
    "    # low_err = [log2_min(sub_norm_elem_rates['gen'][g][t]) - log2_min(max(0.01, sub_norm_elem_rates['low'][g][t])) for t in vis_regions]\n",
    "    # high_err = [log2_min(sub_norm_elem_rates['high'][g][t]) - log2_min(sub_norm_elem_rates['gen'][g][t]) for t in vis_regions]\n",
    "    x = [x * (len(vis_genotypes) + 1) + vis_genotypes.index(g) for x in range(len(vis_regions))]\n",
    "    y = [y - 1 for y in sub_norm_elem_rates['gen'][g].values()]\n",
    "    bars.append(ax.bar(x, y, color=gen_cmap(g), alpha=0.6, bottom=1, yerr=sems[g].values()))\n",
    "for s in vis_samples:\n",
    "    g = s_to_g(s)\n",
    "    x = [x * (len(vis_genotypes) + 1) + vis_genotypes.index(g) + random.random() * 0.5 - 0.25 for x in range(len(vis_regions))]\n",
    "    y = sub_norm_elem_rates['ind'][s].values()\n",
    "    ax.scatter(x, y, color=gen_cmap(g), s=4)\n",
    "# ax.set_ylim(0.37, 2.9)\n",
    "ax.set_yscale('log', base=2)\n",
    "xlim = ax.get_xlim()\n",
    "ax.hlines(1, -10, 100, color='black')\n",
    "ax.set_xlim(xlim)\n",
    "ax.set_xlim(ax.get_xlim()[0] - 0.5, ax.get_xlim()[1] + 0.5)\n",
    "\n",
    "ax.set_xticks([i * (len(vis_genotypes) + 1) + (len(vis_genotypes) - 1) / 2 for i in range(len(vis_regions))])\n",
    "ax.set_xticklabels(vis_regions, rotation=45)\n",
    "p_labels = [f'p={p:.2f}' if p > 0.001 else f'p<e{math.floor(math.log10(max(10**-20, p)))}' for p in p_vals]\n",
    "ax.legend(bars, vis_genotypes, ncol=1, loc='upper left', bbox_to_anchor=(0, 1))\n",
    "ax.set_ylabel('Log2fold mutation rate')\n",
    "# ax.set_title('Mutation rate by genomic region')\n",
    "fig.savefig('figs/mutation_rate_by_element_type_log_fold.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig('figs/mutation_rate_by_element_type_log_fold.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(0, len(vis_regions) - 1, 2):\n",
    "    for g in vis_genotypes:\n",
    "        rate1 = norm_elem_rates['gen'][g][vis_regions[i]]\n",
    "        rate2 = norm_elem_rates['gen'][g][vis_regions[i + 1]]\n",
    "        print(f'{g} {vis_regions[i + 1]}/{vis_regions[i]} {rate2 / rate1:.4f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pvals = []\n",
    "paired_regions = True # whether vis_regions represent one or more pairs of regions to compare (e.g. these regions would be paired: mCG, umCG, mCHG, umCHG)\n",
    "\n",
    "\n",
    "# for each region, compare col-0 to each other group with >1 sample\n",
    "if not paired_regions: # compare col-0 bar to other genotype bar for each region and genotype FIXME put not back\n",
    "    for region in vis_regions:\n",
    "        for g in vis_genotypes:\n",
    "            if g == 'untreated':\n",
    "                continue\n",
    "            if len(g_to_s(g)) > 1:\n",
    "                # welch's two sample t-test\n",
    "                p = stats.ttest_ind([norm_elem_rates['ind'][s][region] for s in g_to_s('untreated')], [norm_elem_rates['ind'][s][region] for s in g_to_s(g)], equal_var=False).pvalue\n",
    "\n",
    "                ratio = norm_elem_rates['gen']['untreated'][region] / norm_elem_rates['gen'][g][region]\n",
    "                pvals.append((p, f'{region} untreated  to {g} two-sample welchs ttest ({ratio:.4f}x or {1/ratio:.4f}x)'))\n",
    "else: # compare ratio of bar1:bar2 for col-0 and other genotypes, then do the same for bar3:bar4...\n",
    "    for i in range(0, len(vis_regions) - 1, 2):\n",
    "        for g in vis_genotypes:\n",
    "            if g == 'untreated' or len(g_to_s(g)) == 1:\n",
    "                continue\n",
    "            r1 = vis_regions[i]\n",
    "            r2 = vis_regions[i + 1]\n",
    "            p = stats.ttest_1samp([norm_elem_rates['ind'][s][r2] / norm_elem_rates['ind'][s][r1] for s in g_to_s('untreated')], norm_elem_rates['gen'][g][r2] / norm_elem_rates['gen'][g][r1]).pvalue\n",
    "            ratios = (norm_elem_rates['gen']['untreated'][r2] / norm_elem_rates['gen']['untreated'][r1],  norm_elem_rates['gen'][g][r2] / norm_elem_rates['gen'][g][r1])\n",
    "            pvals.append((p, f'{r1}/{r2} ratio differs between untreated to {g} one-sample ttest ({ratios[0]:.4f}x to {ratios[1]:.4f}x)'))\n",
    "            \n",
    "# compare pairs of regions for col-0 and ma lines\n",
    "\n",
    "for i, r1 in enumerate(vis_regions[:-1]):\n",
    "    if paired_regions and i % 2 == 1:\n",
    "        continue\n",
    "    comp_regions = [vis_regions[i + 1]] if paired_regions else vis_regions[i + 1:]\n",
    "    for r2 in comp_regions:\n",
    "        p = stats.ttest_ind([norm_elem_rates['ind'][s][r1] for s in g_to_s('untreated')], [norm_elem_rates['ind'][s][r2] for s in g_to_s('untreated')], equal_var=True).pvalue\n",
    "        ratio = norm_elem_rates['gen']['untreated'][r1] / norm_elem_rates['gen']['untreated'][r2]\n",
    "        pvals.append((p, f'untreated {r1} to {r2} two-sample ttest ({ratio:.4f}x or {1/ratio:.4f}x)'))\n",
    "        \n",
    "        p = stats.ttest_ind([norm_elem_rates['ind'][s][r1] for s in g_to_s('MA lines')], [norm_elem_rates['ind'][s][r2] for s in g_to_s('MA lines')], equal_var=True).pvalue\n",
    "        ratio = norm_elem_rates['gen']['MA lines'][r1] / norm_elem_rates['gen']['MA lines'][r2]\n",
    "        pvals.append((p, f'MA lines {r1} to {r2} two-sample ttest ({ratio:.4f}x or {1/ratio:.4f}x)'))\n",
    "\n",
    "# # compare col-0 genes to MA line genes\n",
    "# p = stats.ttest_ind([norm_elem_rates['ind'][s]['gene'] for s in g_to_s('untreated')], [norm_elem_rates['ind'][s]['gene'] for s in g_to_s('MA lines')], equal_var=False).pvalue\n",
    "# pvals.append((p, f'genes Col-0 to MA lines two-sample Welchs ttest'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cutoff_reached = False\n",
    "for i, pval in enumerate(sorted(pvals)):\n",
    "    corrected_p = pval[0] * (len(pvals) - i)\n",
    "    if corrected_p > 0.05 and not cutoff_reached:\n",
    "        print('--------------- significance cutoff -----------------')\n",
    "        cutoff_reached = True\n",
    "    print(corrected_p, pval[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# same as the above plot, but use the raw mutation rate rather than the relative rate\n",
    "fig, ax = plt.subplots(figsize=(len(vis_regions) * (len(vis_genotypes) + 1) * 0.5, 3))\n",
    "bars = []\n",
    "for g in vis_genotypes:\n",
    "    # low_err = [log2_min(sub_norm_elem_rates['gen'][g][t]) - log2_min(max(0.01, sub_norm_elem_rates['low'][g][t])) for t in vis_regions]\n",
    "    # high_err = [log2_min(sub_norm_elem_rates['high'][g][t]) - log2_min(sub_norm_elem_rates['gen'][g][t]) for t in vis_regions]\n",
    "    x = [x * (len(vis_genotypes) + 1) + vis_genotypes.index(g) for x in range(len(vis_regions))]\n",
    "    y = [sub_norm_elem_rates['gen'][g][r] * elem_rates['gen_uniform'][g] for r in vis_regions]\n",
    "    errors = [sems[g][r] * elem_rates['gen_uniform'][g] for r in vis_regions]\n",
    "    bars.append(ax.bar(x, y, color=gen_cmap(g), alpha=0.6, bottom=0, yerr=errors))\n",
    "for s in vis_samples:\n",
    "    g = s_to_g(s)\n",
    "    x = [x * (len(vis_genotypes) + 1) + vis_genotypes.index(g) + random.random() * 0.5 - 0.25 for x in range(len(vis_regions))]\n",
    "    y = [sub_norm_elem_rates['ind'][s][r] * elem_rates['gen_uniform'][g] for r in vis_regions]\n",
    "    ax.scatter(x, y, color=gen_cmap(g), s=4)\n",
    "# ax.set_yscale('log', base=2)\n",
    "xlim = ax.get_xlim()\n",
    "# ax.hlines(1, -10, 100, color='black')\n",
    "ax.set_xlim(xlim)\n",
    "ax.set_xlim(ax.get_xlim()[0] - 0.5, ax.get_xlim()[1] + 0.5)\n",
    "# ax.set_ylim(-1.5, 1.5)\n",
    "\n",
    "ax.set_xticks([i * (len(vis_genotypes) + 1) + (len(vis_genotypes) - 1) / 2 for i in range(len(vis_regions))])\n",
    "ax.set_xticklabels(vis_regions, rotation=45)\n",
    "p_labels = [f'p={p:.2f}' if p > 0.001 else f'p<e{math.floor(math.log10(max(10**-20, p)))}' for p in p_vals]\n",
    "ax.legend(bars, vis_genotypes, ncol=1, loc='upper left', bbox_to_anchor=(0, 1))\n",
    "ax.set_ylabel('mutation rate')\n",
    "# ax.set_title('Mutation rate by genomic region')\n",
    "fig.savefig('figs/mutation_rate_by_element_type_raw.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Mutation rate over chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "yc_mask = {chrom:np.zeros(chrom_sizes[chrom], dtype=bool) for chrom in chrom_sizes}\n",
    "for chrom in chrom_sizes:\n",
    "    for i in tqdm(range(1, len(genome[chrom]) - 1)):\n",
    "        first = genome[chrom][i]\n",
    "        second = genome[chrom][i + 1]\n",
    "        if first == 'G' and (second == 'G' or second == 'A'):\n",
    "            yc_mask[chrom][i] = 1\n",
    "        elif second == 'C' and (first == 'C' or first == 'T'):\n",
    "            yc_mask[chrom][i + 1] = 1\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "window_size = 2000000 # use these for the line plots\n",
    "step_size = 200000\n",
    "\n",
    "# window_size = 1000000 # use these for the scatter plots\n",
    "# step_size = 1000000\n",
    "\n",
    "# count up the amount of mutations and callable coverage in each bin of the genome for each genotype\n",
    "chrom_muts = {g:{chrom:np.zeros(chrom_sizes[chrom] // step_size + 1, dtype=int) for chrom in chrom_sizes} for g in genotypes}\n",
    "# chrom_muts_yc = {g:{chrom:np.zeros(chrom_sizes[chrom] // step_size + 1, dtype=int) for chrom in chrom_sizes} for g in genotypes}\n",
    "chrom_covs = {g:{chrom:np.zeros(chrom_sizes[chrom] //  step_size + 1, dtype=int) for chrom in chrom_sizes} for g in genotypes}\n",
    "# chrom_covs_yc = {g:{chrom:np.zeros(chrom_sizes[chrom] //  step_size + 1, dtype=int) for chrom in chrom_sizes} for g in genotypes}\n",
    "total_chrom_cov = {g:{chrom:0 for chrom in chrom_sizes} for g in genotypes}\n",
    "for chrom in chrom_sizes:\n",
    "    ma_cov_arr = np.load(f'{cov_prefixes[\"untreated_1\"]}{chrom}.npy') > 0\n",
    "    # ma_cov_arr_yc = np.multiply(yc_mask[chrom], ma_cov_arr)\n",
    "    \n",
    "    for sample in tqdm(sample_names):\n",
    "        # count mutations\n",
    "        df = dfs_somatic[sample]\n",
    "        genotype = s_to_g(sample)\n",
    "        mut_arr = np.zeros(chrom_sizes[chrom], dtype=int)\n",
    "        for pos in df[df.chrom == chrom].pos:\n",
    "            mut_arr[pos] += 1\n",
    "        # mut_arr_yc = np.multiply(yc_mask[chrom], mut_arr)\n",
    "            \n",
    "        # count callable coverage\n",
    "        if 'Ler' in genotype: # use sites with >0 coverage\n",
    "            cov_arr = np.load(f'{cov_prefixes[sample]}{chrom}.npy')\n",
    "            cov_arr = cov_arr > 0\n",
    "            # cov_arr_yc = np.multiply(yc_mask[chrom], cov_arr)\n",
    "        elif '1001' in genotype or 'MA lines' in genotype: # use sites in Col-0-1 with >0 coverage\n",
    "            cov_arr = ma_cov_arr\n",
    "            # cov_arr_yc = ma_cov_arr_yc\n",
    "        else:\n",
    "            cov_arr = np.load(f'{cov_prefixes[sample]}{chrom}.npy')\n",
    "            # cov_arr_yc = np.multiply(yc_mask[chrom], cov_arr)\n",
    "        \n",
    "        for i, center in enumerate(range(0, chrom_sizes[chrom], step_size)):\n",
    "            chrom_muts[genotype][chrom][i] += np.sum(mut_arr[max(0, center - window_size // 2):center + window_size // 2])\n",
    "            # chrom_muts_yc[genotype][chrom][i] += np.sum(mut_arr_yc[max(0, center - window_size // 2):center + window_size // 2])\n",
    "            chrom_covs[genotype][chrom][i] += np.sum(cov_arr[max(0, center - window_size // 2):center + window_size // 2])\n",
    "            # chrom_covs_yc[genotype][chrom][i] += np.sum(cov_arr_yc[max(0, center - window_size // 2):center + window_size // 2])\n",
    "        total_chrom_cov[genotype][chrom] += np.sum(cov_arr)\n",
    "\n",
    "chrom_boot_lower = {chrom:np.zeros(chrom_sizes[chrom] // step_size + 1, dtype=float) for chrom in chrom_sizes}\n",
    "chrom_boot_upper = {chrom:np.zeros(chrom_sizes[chrom] // step_size + 1, dtype=float) for chrom in chrom_sizes}\n",
    "for chrom in chrom_sizes:\n",
    "    # get the number of mutations and callable coverage per bin for each of the 8 WT replicates\n",
    "    mut_arrs = [np.zeros(chrom_sizes[chrom] // step_size + 1, dtype=int) for x in range(8)]\n",
    "    cov_arrs = [np.zeros(chrom_sizes[chrom] // step_size + 1, dtype=int) for x in range(8)]\n",
    "    for i in range(8):\n",
    "        mut_arr = np.zeros(chrom_sizes[chrom], dtype=int)\n",
    "        cov_arr = np.zeros(chrom_sizes[chrom], dtype=int)\n",
    "        df = dfs_somatic[f'untreated_{i+1}']\n",
    "        for pos in df[df.chrom == chrom].pos:\n",
    "            mut_arr[pos] += 1\n",
    "        cov_arr = np.load(f'{cov_prefixes[\"untreated_\" + str(i+1)]}{chrom}.npy')\n",
    "        \n",
    "        for bin_idx, center in enumerate(range(0, chrom_sizes[chrom], step_size)):\n",
    "            mut_arrs[i][bin_idx] = np.sum(mut_arr[max(0, center - window_size // 2):center + window_size // 2])\n",
    "            cov_arrs[i][bin_idx] = np.sum(cov_arr[max(0, center - window_size // 2):center + window_size // 2])\n",
    "            \n",
    "    # for 2000 bootstraps of 8 samples, get a mutation rate\n",
    "    boot_rates = np.zeros((2000, len(mut_arrs[0])), dtype=float)\n",
    "    for b in range(2000):\n",
    "        boot_samples = random.choices(range(8), k=8)\n",
    "        mut_arr = np.sum([mut_arrs[x] for x in boot_samples], axis=0)\n",
    "        cov_arr = np.sum([cov_arrs[x] for x in boot_samples], axis=0)\n",
    "        boot_rates[b] = np.divide(mut_arr, cov_arr)\n",
    "    \n",
    "    lower_boots = np.percentile(boot_rates, 2.5, axis=0)\n",
    "    upper_boots = np.percentile(boot_rates, 97.5, axis=0)\n",
    "    assert len(lower_boots) == len(mut_arrs[0])\n",
    "    chrom_boot_lower[chrom] = lower_boots\n",
    "    chrom_boot_upper[chrom] = upper_boots\n",
    "    \n",
    "        \n",
    "# get an upper and lower bound for mutation number assuming binomial distribution\n",
    "chrom_muts_lower = {g:{chrom:np.zeros(chrom_sizes[chrom] // step_size + 1, dtype=int) for chrom in chrom_sizes} for g in genotypes}\n",
    "chrom_muts_upper = {g:{chrom:np.zeros(chrom_sizes[chrom] // step_size + 1, dtype=int) for chrom in chrom_sizes} for g in genotypes}\n",
    "for g in genotypes:\n",
    "    total_muts = sum([sum(chrom_muts[g][chrom]) for chrom in chrom_sizes])\n",
    "    for chrom in chrom_sizes:\n",
    "        for i in range(len(chrom_muts[g][chrom])):\n",
    "            lower, upper = stats.binom.interval(0.95, total_muts, max(1, chrom_muts[g][chrom][i]) / total_muts)\n",
    "            chrom_muts_lower[g][chrom][i] = lower\n",
    "            chrom_muts_upper[g][chrom][i] = upper\n",
    "            \n",
    "chrom_rates = dict()\n",
    "chrom_rates_yc = dict()\n",
    "chrom_rates_lower = dict()\n",
    "chrom_rates_upper = dict()\n",
    "chrom_rates_boot_lower = dict()\n",
    "chrom_rates_boot_upper = dict()\n",
    "for g in genotypes:\n",
    "    total_muts = sum([sum(chrom_muts[g][chrom]) for chrom in chrom_sizes])\n",
    "    # total_muts_yc = sum([sum(chrom_muts[g][chrom]) for chrom in chrom_sizes])\n",
    "    total_cov = sum([sum(chrom_covs[g][chrom]) for chrom in chrom_sizes])\n",
    "    # total_cov_yc = sum([sum(chrom_covs_yc[g][chrom]) for chrom in chrom_sizes])\n",
    "    chrom_rates[g] = {chrom:(chrom_muts[g][chrom] / (chrom_covs[g][chrom] + 1)) / (total_muts / total_cov) for chrom in chrom_sizes}\n",
    "    # chrom_rates_yc[g] = {chrom:(chrom_muts_yc[g][chrom] / (chrom_covs_yc[g][chrom] + 1)) / (total_muts_yc / total_cov_yc) for chrom in chrom_sizes}\n",
    "    chrom_rates_lower[g] = {chrom:(chrom_muts_lower[g][chrom] / (chrom_covs[g][chrom] + 1)) / (total_muts / total_cov) for chrom in chrom_sizes}\n",
    "    chrom_rates_upper[g] = {chrom:(chrom_muts_upper[g][chrom] / (chrom_covs[g][chrom] + 1)) / (total_muts / total_cov) for chrom in chrom_sizes}\n",
    "    \n",
    "    if g == 'untreated':\n",
    "        chrom_rates_boot_lower = {chrom:chrom_boot_lower[chrom] / (total_muts / total_cov) for chrom in chrom_sizes}\n",
    "        chrom_rates_boot_upper = {chrom:chrom_boot_upper[chrom] / (total_muts / total_cov) for chrom in chrom_sizes}\n",
    "        print(total_muts,  total_cov)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vis_genotypes = genotypes\n",
    "vis_samples = [s for s in sample_names if s_to_g(s) in vis_genotypes]\n",
    "\n",
    "import matplotlib.lines as mlines\n",
    "\n",
    "# centromere_pos = {'Chr1':16029005, 'Chr2':5731747, 'Chr3':14711236, 'Chr4':6595103, 'Chr5':13790251} # don't remember what paper this is from, but it's supposed to be the center of the centromere repeats\n",
    "centromere_pos = {'Chr1':15256434, 'Chr2':3697368, 'Chr3':13599663, 'Chr4':3941890, 'Chr5':11880011} # this is from the Col-CEN paper (34762468) table S1 and figure S3, center of the centromere\n",
    "fig, axs = plt.subplots(len(chrom_sizes), figsize=(10, 6), sharex=True)\n",
    "for i, chrom in enumerate(chrom_sizes):\n",
    "    lines = []\n",
    "    for genotype in vis_genotypes:\n",
    "        l = axs[i].plot([x * step_size for x in range(len(chrom_rates[genotype][chrom]))], chrom_rates[genotype][chrom], c=gen_cmap(genotype))\n",
    "        if genotype == 'untreated':\n",
    "            # axs[i].fill_between([x * step_size for x in range(len(chrom_rates[genotype][chrom]))], chrom_rates_lower[genotype][chrom], chrom_rates_upper[genotype][chrom], color=gen_cmap(genotype), alpha=0.2, edgecolors='none')\n",
    "            axs[i].fill_between([x * step_size for x in range(len(chrom_rates[genotype][chrom]))], chrom_rates_boot_lower[chrom], chrom_rates_boot_upper[chrom], color=gen_cmap(genotype), alpha=0.2, edgecolors='none')\n",
    "        lines.append(l[0])\n",
    "    ylim = axs[i].get_ylim()\n",
    "    axs[i].vlines(centromere_pos[chrom], -1, axs[i].get_ylim()[1] + 1, color='black')\n",
    "    axs[i].set_ylim(ylim)\n",
    "    axs[i].set_title(chrom, y=0.6, x=0.95)\n",
    "axs[-1].set_xlabel('position (bp)')\n",
    "axs[0].set_xlim(0, 30000000)\n",
    "axs[2].set_ylabel('Mutation/polymorphism rate relative to mean')\n",
    "axs[0].legend(lines, vis_genotypes, loc='upper left', bbox_to_anchor=(1, 1))\n",
    "# axs[0].legend(['untreated', '95% confidence (mutuations per bin ~ binomial)', '95% confidence (2000 bootstraps of 8 samples)'], bbox_to_anchor=(1,1))\n",
    "fig.suptitle('Mutation rate across chromosomes')\n",
    "fig.savefig('figs/mutation_rate_across_chromosomes.png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.lines as mlines\n",
    "\n",
    "# centromere_pos = {'Chr1':16029005, 'Chr2':5731747, 'Chr3':14711236, 'Chr4':6595103, 'Chr5':13790251} # don't remember what paper this is from, but it's supposed to be the center of the centromere repeats\n",
    "centromere_pos = {'Chr1':15256434, 'Chr2':3697368, 'Chr3':13599663, 'Chr4':3941890, 'Chr5':11880011} # this is from the Col-CEN paper (34762468) table S1 and figure S3, center of the centromere\n",
    "fig, axs = plt.subplots(len(chrom_sizes), figsize=(10, 6), sharex=True)\n",
    "for i, chrom in enumerate(chrom_sizes):\n",
    "    lines = []\n",
    "    for genotype in vis_genotypes:\n",
    "        l = axs[i].plot([x * step_size for x in range(len(chrom_rates_yc[genotype][chrom]))], chrom_rates_yc[genotype][chrom], c=gen_cmap(genotype))\n",
    "        lines.append(l[0])\n",
    "    ylim = axs[i].get_ylim()\n",
    "    axs[i].vlines(centromere_pos[chrom], -1, axs[i].get_ylim()[1] + 1, color='black')\n",
    "    axs[i].set_ylim(ylim)\n",
    "    axs[i].set_title(chrom, y=0.6, x=0.95)\n",
    "axs[-1].set_xlabel('position (bp)')\n",
    "axs[0].set_xlim(0, 30000000)\n",
    "axs[2].set_ylabel('Relative mutation rate of YC sites')\n",
    "axs[0].legend(lines, vis_genotypes, loc='upper left', bbox_to_anchor=(1, 1))\n",
    "# axs[0].legend(['untreated', '95% confidence (mutuations per bin ~ binomial)', '95% confidence (2000 bootstraps of 8 samples)'], bbox_to_anchor=(1,1))\n",
    "fig.suptitle('Mutation rate of YC sites across chromosomes')\n",
    "fig.savefig('figs/mutation_rate_across_chromosomes_yc.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# centromere_pos = {'Chr1':16029005, 'Chr2':5731747, 'Chr3':14711236, 'Chr4':6595103, 'Chr5':13790251} # don't remember what paper this is from, but it's supposed to be the center of the centromere repeats\n",
    "centromere_pos = {'Chr1':15256434, 'Chr2':3697368, 'Chr3':13599663, 'Chr4':3941890, 'Chr5':11880011} # this is from the Col-CEN paper (34762468) table S1 and figure S3, center of the centromere\n",
    "fig, axs = plt.subplots(len(chrom_sizes), figsize=(10, 6), sharex=True, sharey=True)\n",
    "for i, chrom in enumerate(chrom_sizes):\n",
    "    lines = []\n",
    "    for genotype in vis_genotypes:\n",
    "        x = [x * step_size for x in range(len(chrom_covs[genotype][chrom]))]\n",
    "        x_sizes = np.array([min(chrom_sizes[chrom], center + window_size // 2) - max(0, center - window_size // 2) for center in x])\n",
    "        # print(chrom_covs[genotype][chrom] / x_sizes)\n",
    "        # print(total_chrom_cov[genotype] / chrom_sizes[chrom])\n",
    "        # print(chrom_covs[genotype][chrom])\n",
    "        y = (chrom_covs[genotype][chrom] / x_sizes) / (total_chrom_cov[genotype][chrom] / chrom_sizes[chrom]) # coverage per site / genome average\n",
    "        l = axs[i].plot(x, y, c=gen_cmap(genotype))\n",
    "        lines.append(l[0])\n",
    "    ylim = axs[i].get_ylim()\n",
    "    axs[i].vlines(centromere_pos[chrom], -1, axs[i].get_ylim()[1] + 1, color='black')\n",
    "    axs[i].set_ylim(ylim)\n",
    "    axs[i].set_title(chrom, y=0.6, x=0.95)\n",
    "axs[-1].set_xlabel('position (bp)')\n",
    "axs[0].set_xlim(0, 30000000)\n",
    "axs[0].set_ylim(0, axs[0].get_ylim()[1])\n",
    "axs[2].set_ylabel('Callable coverage relative to mean')\n",
    "axs[0].legend(lines, vis_genotypes, loc='upper left', bbox_to_anchor=(1, 1))\n",
    "# axs[0].legend(['untreated', '95% confidence (mutuations per bin ~ binomial)', '95% confidence (2000 bootstraps of 8 samples)'], bbox_to_anchor=(1,1))\n",
    "fig.suptitle('Callable coverage across chromosomes')\n",
    "fig.savefig('figs/callable_coverage_across_chromosomes.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scatterplot best fit lines of the above data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the distance of each bin from the centromere\n",
    "bin_dists = {chrom:[abs(step_size * i - centromere_pos[chrom]) for i in range(chrom_sizes[chrom] // step_size + 1)] for chrom in chrom_sizes}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# visualize correlation of distance from centromere with substitution rate\n",
    "cols = int(len(vis_genotypes) ** (1/2))\n",
    "fig, axs = plt.subplots(len(vis_genotypes) // cols, cols, figsize=(len(vis_genotypes), len(vis_genotypes)), sharex=True)\n",
    "\n",
    "for i, g in enumerate(vis_genotypes):\n",
    "    if cols > 1:\n",
    "        ax = axs[i // cols, i % cols]\n",
    "    else:\n",
    "        ax = axs[i]\n",
    "    \n",
    "    x = []\n",
    "    y = []\n",
    "    for chrom in chrom_sizes:\n",
    "        x += bin_dists[chrom]\n",
    "        y += list(chrom_rates[g][chrom])\n",
    "    \n",
    "    slope, intersect = np.polyfit(x, y, 1)\n",
    "    color = gen_cmap(g)\n",
    "    r = np.corrcoef(x, y)\n",
    "    ax.scatter(x, y, color=color, s=5)\n",
    "    ax.axline((0, intersect), slope=slope, color=color)\n",
    "    ax.set_title(f'{g} r={r[0][1]:.2f}', y=0.8)\n",
    "    ax.set_ylim(0, ax.get_ylim()[1])\n",
    "    ax.set_ylabel('Relative substitution rate')\n",
    "    ax.set_xlabel('Distance from centromere')\n",
    "    \n",
    "fig.suptitle('Substitution rate correlates with distance from centromere')\n",
    "fig.savefig('figs/dist_from_centromere_scatters.svg', bbox_inches='tight', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# YC sites only\n",
    "cols = int(len(vis_genotypes) ** (1/2))\n",
    "fig, axs = plt.subplots(len(vis_genotypes) // cols, cols, figsize=(len(vis_genotypes), len(vis_genotypes)), sharex=True)\n",
    "\n",
    "for i, g in enumerate(vis_genotypes):\n",
    "    ax = axs[i // cols, i % cols]\n",
    "    \n",
    "    x = []\n",
    "    y = []\n",
    "    for chrom in chrom_sizes: #chrom_sizes: # FIXME\n",
    "        x += bin_dists[chrom]\n",
    "        y += list(chrom_rates_yc[g][chrom])\n",
    "    \n",
    "    slope, intersect = np.polyfit(x, y, 1)\n",
    "    color = gen_cmap(g)\n",
    "    r = np.corrcoef(x, y)\n",
    "    ax.scatter(x, y, color=color, s=5)\n",
    "    ax.axline((0, intersect), slope=slope, color=color)\n",
    "    ax.set_title(f'{g} r={r[0][1]:.2f}', y=0.8)\n",
    "    ax.set_ylim(0, ax.get_ylim()[1])\n",
    "    \n",
    "axs[len(axs) // 2][0].set_ylabel('Relative mutation rate of YC sites')\n",
    "axs[-1][cols // 2].set_xlabel('Distance from centromere')\n",
    "fig.suptitle('mutation rate and distance to centromere (YC sites only)')\n",
    "fig.savefig('figs/dist_from_centromere_scatters_yc.png', bbox_inches='tight', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(bin_dists['Chr2']), len(chrom_rates[genotypes[0]]['Chr2'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Mutation hotspots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hotspot_size = 10000\n",
    "\n",
    "# make a histogram of the number of mutations within half-overlapping sliding windows of the genome\n",
    "obs_hist = {g:np.zeros(10000) for g in genotypes}\n",
    "genotype_muts = {g:0 for g in genotypes} # total number of mutations for the genotype\n",
    "for g in tqdm(genotypes):\n",
    "    # make arrays of the mutations per site for the genotype\n",
    "    obs_muts = {chrom:np.zeros(chrom_sizes[chrom], dtype=np.ubyte) for chrom in chrom_sizes}\n",
    "    samples = [s for s in sample_names if s_to_g(s) == g] # samples belonging to genotype\n",
    "    for s in samples:\n",
    "        for row in dfs_somatic[s].itertuples():\n",
    "            obs_muts[row.chrom][row.pos] += 1\n",
    "            genotype_muts[g] += 1\n",
    "    \n",
    "    # iterate over sliding windows, find the mutation count in the window and increment obs_hist\n",
    "    for chrom in chrom_sizes:\n",
    "        i = 0\n",
    "        while i < chrom_sizes[chrom]:\n",
    "            muts = np.sum(obs_muts[chrom][i:i + hotspot_size])\n",
    "            obs_hist[g][muts] += 1\n",
    "            i += hotspot_size // 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculated the expected histogram as a binomial distribution\n",
    "exp_hist = {g:np.zeros(10000) for g in genotypes}\n",
    "for g in genotypes:\n",
    "    for i in range(len(exp_hist[g])):\n",
    "        exp_hist[g][i] = stats.binom.pmf(i, genotype_muts[g], hotspot_size / sum(chrom_sizes.values())) * np.sum(obs_hist[g])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vis_genotypes = ['untreated', 'UVB', 'UVC', 'MA lines']\n",
    "fig, axs = plt.subplots(len(vis_genotypes), figsize=(8, len(vis_genotypes) * 1.5), gridspec_kw={'hspace':0.3}, sharex=True)\n",
    "for i, g in enumerate(vis_genotypes):\n",
    "    max_x = np.max(np.nonzero(obs_hist[g])) + 10\n",
    "    axs[i].hlines(1, -1, max_x + 2, color='black')\n",
    "    axs[i].plot(range(max_x), obs_hist[g][:max_x], '-o', color=gen_cmap(g))\n",
    "    axs[i].plot(range(max_x), exp_hist[g][:max_x], '-o', color='tab:grey')\n",
    "    axs[i].set_title(g, y=0.6)\n",
    "    axs[i].set_yscale('log')\n",
    "    # print(axs[i].get_ylim(), np.max(obs_hist[g][:max_x]), np.max(exp_hist[g][:max_x]))\n",
    "    axs[i].set_ylim(0.01, np.max(exp_hist[g]) * 1.5)\n",
    "    axs[i].set_xlim(-0.3, max_x - 0.7)\n",
    "axs[-1].set_xlabel('number of mutations in window')\n",
    "axs[len(axs) // 2].set_ylabel('count')\n",
    "fig.savefig('figs/hotspots_for_mutations.svg', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Logistic regression\n",
    "\n",
    "this section is very memory heavy, requiring ~400GB when doing regression on the whole genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "del df_gff\n",
    "for r in ['intron', 'exon', 'te', 'promoter', 'intergenic', 'unmet cg', 'met cg', 'unmet chg', 'met chg', 'unmet chh', 'met chh']:\n",
    "    del region_arrays[r]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import statsmodels.api as sm\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# used for converting chrom:pos positions to just a single index of a 1D array\n",
    "chrom_offset = dict()\n",
    "tot = 0\n",
    "for chrom in chrom_sizes:\n",
    "    chrom_offset[chrom] = tot\n",
    "    tot += chrom_sizes[chrom]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_genotype = '1001'\n",
    "reg_region = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cg = np.concatenate(list(region_arrays['cg'].values())) # note that this is not CG vs AT, but CG vs AT/NN/YR/SS/WW\n",
    "peri = np.concatenate(list(region_arrays['pericentromere'].values()))\n",
    "met = np.concatenate(list(region_arrays['met c'].values())) # note that this is not met vs unmet, but met vs unmet/unknown\n",
    "acr = np.concatenate(list(region_arrays['acr'].values()))\n",
    "genic = np.concatenate(list(region_arrays['gene'].values()))\n",
    "ns = np.concatenate(list(ns_sites.values())) # note that this is not nonsyn vs syn, but nonsyn vs syn/nonCDS\n",
    "\n",
    "intercept = np.ones(len(cg), dtype=bool) # statsmodels does not use an intercept by default, so a column of 1s needs to be added to the explanatory variables\n",
    "\n",
    "# make an array of predictor variables (reg_x)\n",
    "if 'Ler-0' in reg_genotype or '1001' in reg_genotype:\n",
    "    reg_x = np.vstack((intercept, cg, peri, met, acr, genic, acr * genic, ns)).T # explanatory variables, each column is one variable (e.g. methylation)\n",
    "    predictors = ['intercept', 'cg', 'peri', 'met', 'acr', 'genic', 'acr*genic', 'ns']\n",
    "else:\n",
    "    reg_x = np.vstack((intercept, cg, peri, met, acr, genic, acr * genic)).T # explanatory variables, each column is one variable (e.g. methylation)\n",
    "    predictors = ['intercept', 'cg', 'peri', 'met', 'acr', 'genic', 'acr*genic']\n",
    "\n",
    "del cg, peri, met, acr, ns, intercept, genic\n",
    "\n",
    "\n",
    "# count the number of mutations at each site in the genome (i.e. the successes of a binomial)\n",
    "reg_successes = np.zeros(len(reg_x), dtype=int)\n",
    "for s in sample_names:\n",
    "    if s_to_g(s) != reg_genotype:\n",
    "        continue\n",
    "    for r in dfs_somatic[s].itertuples():\n",
    "        reg_successes[chrom_offset[r.chrom] + r.pos] += 1\n",
    "\n",
    "# count the amount of coverage at each site in the genome (i.e. the trials of a binomial)\n",
    "if 'Ler' in reg_genotype: # get all sites with at least one coverage in ler\n",
    "    arr = np.zeros(0)\n",
    "    for chrom in chrom_sizes:\n",
    "        arr = np.append(arr, np.load(f'{cov_prefixes[\"Ler-0\"]}{chrom}.npy'))\n",
    "    arr = arr > 0\n",
    "elif '1001' in reg_genotype or 'MA lines' in reg_genotype: # get all sites with at least one coverage in Col-0-1\n",
    "    arr = np.zeros(0)\n",
    "    for chrom in chrom_sizes:\n",
    "        arr = np.append(arr, np.load(f'{cov_prefixes[\"untreated_1\"]}{chrom}.npy'))\n",
    "    arr = arr > 0\n",
    "else: # get the sum of callable coverage across samples\n",
    "    arr = np.zeros(sum(chrom_sizes.values()))\n",
    "    for s in sample_names:\n",
    "        if s_to_g(s) != reg_genotype:\n",
    "            continue\n",
    "        for chrom in chrom_sizes:\n",
    "            arr[chrom_offset[chrom]:chrom_offset[chrom] + chrom_sizes[chrom]] += np.load(f'{cov_prefixes[s]}{chrom}.npy')\n",
    "reg_trials = arr\n",
    "\n",
    "# restrict x and y to one region (e.g. run the model on promoters only)\n",
    "if reg_region is not None:\n",
    "    restrict = np.concatenate(list(region_arrays[reg_region].values()))\n",
    "    reg_x = reg_x[restrict]\n",
    "    reg_successes = reg_successes[restrict]\n",
    "    reg_trials = reg_trials[restrict]\n",
    "print(len(reg_x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# only use sites where there was callable coverage (the model cannot handle nan values)\n",
    "x_subset = reg_x[reg_trials > 0]\n",
    "success_subset = reg_successes[reg_trials > 0]\n",
    "trial_subset = reg_trials[reg_trials > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'MA lines' in reg_genotype or 'Ler-0' in reg_genotype:\n",
    "    y_subset = np.clip(success_subset, None, 1).T # only allow one mutation per site\n",
    "    model = sm.Logit(y_subset, x_subset) # use logistic regression, as each site has a chance of containing a mutation or not\n",
    "elif '1001' in reg_genotype:\n",
    "    y_subset = success_subset.T\n",
    "    model = sm.Poisson(y_subset, x_subset) # use poisson regression, as each site can have different mutations in different accessions (not a great model, as we can't distinguish double hits)\n",
    "else:\n",
    "    y_subset = np.vstack((success_subset, trial_subset - success_subset)).T\n",
    "    model = sm.GLM(y_subset, x_subset, family=sm.families.Binomial()) # usually takes 15 mins on the whole genome\n",
    "    \n",
    "res = model.fit()\n",
    "print(res.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the effect size of each explanatory variable\n",
    "means = np.mean(x_subset, axis=0) # get mean of each column (fraction of genome with each flag)\n",
    "for i, var in enumerate(predictors):\n",
    "    arr = np.vstack((means, means)) # it's not necessary to use the means when gettingthe ratio of p[0]:p[1]\n",
    "    arr[0][i] = 0\n",
    "    arr[1][i] = 1\n",
    "    p = res.predict(arr)\n",
    "    print(f'predicted effect of {var}: {p[0]:.3e} -> {p[1]:.3e} ({p[1] / p[0]:.3f}x)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res.save(f'tmp/regression_{reg_genotype}_{reg_region}_{\"\".join([x[0] for x in predictors])}.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.iolib.smpickle import load_pickle\n",
    "models = {'untreated':load_pickle('tmp/regression_untreated_None_icpmaga.pkl'),\n",
    "          'MA lines':load_pickle('tmp/regression_MA lines_None_icpmaga.pkl'),\n",
    "          'Ler-0':load_pickle('tmp/regression_Ler-0_None_icpmagan.pkl'),\n",
    "          '1001':load_pickle('tmp/regression_1001_None_icpmagan.pkl')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.e**-17.7484"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "models['untreated'].conf_int()\n",
    "models['untreated'].params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for model in models:\n",
    "#     print(model)\n",
    "#     print(models[model].summary())\n",
    "#     row = ''\n",
    "#     for i in range(len(models[model].params)):\n",
    "#         coef = models[model].params[i]\n",
    "#         interval = models[model].conf_int()[i]\n",
    "#         if coef < -2:\n",
    "#             row += f'{np.e ** coef:.2e} [{np.e ** interval[0]:.2e}, {np.e ** interval[1]:.2e}]\\n'\n",
    "#         else:\n",
    "#             row += f'{np.e ** coef:.2f} [{np.e ** interval[0]:.2f}, {np.e ** interval[1]:.2f}]\\n'\n",
    "            \n",
    "#     print(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "avg_rates = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate the genome wide average rate for the model. used to calculate relative rate\n",
    "# this code block needs to be run multiple times after updating reg_genotype and x_subset above\n",
    "# I'm not super sure why these don't match gen_rates, they do for MA lines and 1001 but not Ler-0\n",
    "pred_vals = models[reg_genotype].predict(x_subset)\n",
    "avg_rate = np.dot(pred_vals, trial_subset) / np.sum(trial_subset)\n",
    "avg_rates[reg_genotype] = avg_rate\n",
    "print(reg_genotype, avg_rate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "avg_rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acr_genic = dict()\n",
    "acr_nongenic = dict()\n",
    "nonacr_genic = dict()\n",
    "nonacr_nongenic = dict()\n",
    "\n",
    "means = np.mean(x_subset, axis=0)\n",
    "for g in models:\n",
    "    x = means\n",
    "    if '1001' in g or 'Ler' in g:\n",
    "        x = means.copy()\n",
    "    else:\n",
    "        x = means[:-1].copy()\n",
    "    \n",
    "    covar = models[g].cov_params()\n",
    "    \n",
    "    avg_rate = avg_rates[g]\n",
    "    \n",
    "    # genic ACR rate\n",
    "    x[4] = 1\n",
    "    x[5] = 1\n",
    "    x[6] = 1\n",
    "    pred = models[g].predict(x) # predict the mutation rate for genic ACRs\n",
    "    gradient = (pred * (1-pred) * x.T).T # I don't really know how this math works, but it's using the delta method to estimate a confidence interval assuming normal dist\n",
    "    std_err = np.sqrt(np.dot(np.dot(gradient, covar), gradient))\n",
    "    # acr_genic[g] = [pred[0] / avg_rate, 1.96 * std_err / avg_rate]\n",
    "    acr_genic[g] = [math.log2(pred[0] / avg_rate), math.log2(pred[0] / avg_rate) - math.log2(pred[0] / avg_rate - 1.0 * std_err / avg_rate)]\n",
    "    \n",
    "    # nongenic ACR rate\n",
    "    x[4] = 1\n",
    "    x[5] = 0\n",
    "    x[6] = 0\n",
    "    pred = models[g].predict(x)\n",
    "    gradient = (pred * (1-pred) * x.T).T\n",
    "    std_err = np.sqrt(np.dot(np.dot(gradient, covar), gradient))\n",
    "    # acr_nongenic[g] = [pred[0] / avg_rate, 1.96 * std_err / avg_rate]\n",
    "    acr_nongenic[g] = [math.log2(pred[0] / avg_rate), math.log2(pred[0] / avg_rate) - math.log2(pred[0] / avg_rate - 1.0 * std_err / avg_rate)]\n",
    "    \n",
    "    # genic nonACR rate\n",
    "    x[4] = 0\n",
    "    x[5] = 1\n",
    "    x[6] = 0\n",
    "    pred = models[g].predict(x)\n",
    "    gradient = (pred * (1-pred) * x.T).T\n",
    "    std_err = np.sqrt(np.dot(np.dot(gradient, covar), gradient))\n",
    "    # nonacr_genic[g] = [pred[0] / avg_rate, 1.96 * std_err / avg_rate]\n",
    "    nonacr_genic[g] = [math.log2(pred[0] / avg_rate), math.log2(pred[0] / avg_rate) - math.log2(pred[0] / avg_rate - 1.0 * std_err / avg_rate)]\n",
    "    \n",
    "    # nongenic nonACR rate\n",
    "    x[4] = 0\n",
    "    x[5] = 0\n",
    "    x[6] = 0\n",
    "    pred = models[g].predict(x)\n",
    "    gradient = (pred * (1-pred) * x.T).T\n",
    "    std_err = np.sqrt(np.dot(np.dot(gradient, covar), gradient))\n",
    "    # nonacr_nongenic[g] = [pred[0] / avg_rate, 1.96 * std_err / avg_rate]\n",
    "    nonacr_nongenic[g] = [math.log2(pred[0] / avg_rate), math.log2(pred[0] / avg_rate) - math.log2(pred[0] / avg_rate - 1.0 * std_err / avg_rate)]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(6, 4))\n",
    "bars = []\n",
    "for i, g in enumerate(models):\n",
    "    x = [0 + 0.1 * (i - 2), 1 + 0.1 * (i - 2)]\n",
    "    y = [acr_genic[g][0], nonacr_genic[g][0]]\n",
    "    bar = ax.errorbar(x, y, fmt='o-', yerr=[[acr_genic[g][1], nonacr_genic[g][1]], [acr_genic[g][1], nonacr_genic[g][1]]], color=gen_cmap(g))\n",
    "    bars.append(bar)\n",
    "    \n",
    "    x = [2 + 0.1 * (i - 2), 3 + 0.1 * (i - 2)]\n",
    "    y = [acr_nongenic[g][0], nonacr_nongenic[g][0]]\n",
    "    ax.errorbar(x, y, fmt='o-', yerr=[[acr_nongenic[g][1], nonacr_nongenic[g][1]], [acr_nongenic[g][1], nonacr_nongenic[g][1]]], color=gen_cmap(g))\n",
    "\n",
    "ax.hlines(0, -1, 5, color='black')\n",
    "ax.set_xlim(-0.3, 3.3)\n",
    "ax.set_xticks(range(4), labels=['ACR genic', 'nonACR genic', 'ACR nongenic', 'nonACR nongenic'])\n",
    "ax.set_ylabel('log2fold relative substitution rate')\n",
    "ax.legend(bars, models)\n",
    "fig.savefig('figs/regression_acr_effect.svg', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acr_genic, nonacr_genic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_chrom_sizes(index):\n",
    "    chrom_sizes = dict()\n",
    "    with open(index, 'r') as f:\n",
    "        for l in f:\n",
    "            split = l[:-1].split('\\t')\n",
    "            chrom_sizes[split[0]] = int(split[1])\n",
    "    return chrom_sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_sizes = load_chrom_sizes('data/ref/ref.fa.fai')\n",
    "chrom_sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cg_sites = {chrom:np.zeros(chrom_sizes[chrom], dtype=bool) for chrom in chrom_sizes}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cg_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/methyl/bewick_bisulfite_1.allc', 'r') as f:\n",
    "    for l in f:\n",
    "        l = l.split('\\t')\n",
    "        chrom = l[0]\n",
    "        pos = l[1]\n",
    "        cg_sites[chrom][int(pos)] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cg_sites['Chr2'][15]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cg_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cgs_near = {chrom:np.zeros(chrom_sizes[chrom], dtype=int) for chrom in chrom_sizes}\n",
    "for chrom in chrom_sizes:\n",
    "    for i in tqdm(range(chrom_sizes[chrom])):\n",
    "        cgs_near[chrom][i] = sum(cg_sites[chrom][i - 50:i + 50])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cgs_near = {chrom:np.zeros(chrom_sizes[chrom], dtype=int) for chrom in chrom_sizes}\n",
    "with open('data/methyl/bewick_bisulfite_1.allc', 'r') as f:\n",
    "    for l in tqdm(f):\n",
    "        l = l.split('\\t')\n",
    "        chrom = l[0]\n",
    "        pos = l[1]\n",
    "        cgs_near[chrom][int(pos) - 50:int(pos) + 50] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cgs_near"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cgs_near['Chr1'] > 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # calculate the mutation rate at CG vs AT sites\n",
    "# samples = [s for s in sample_names if s_to_g(s) == 'Col-0']\n",
    "# cg_muts = 0\n",
    "# at_muts = 0\n",
    "# for s in samples:\n",
    "#     cg_muts += sum(dfs_somatic[s].ref == 'C') + sum(dfs_somatic[s].ref == 'G')\n",
    "#     at_muts += sum(dfs_somatic[s].ref == 'A') + sum(dfs_somatic[s].ref == 'T')\n",
    "\n",
    "# cg_cov = 0\n",
    "# at_cov = 0\n",
    "# for chrom in tqdm(chrom_sizes):\n",
    "#     cg_arr = np.zeros(chrom_sizes[chrom], dtype=bool)\n",
    "#     at_arr = np.zeros(chrom_sizes[chrom], dtype=bool)\n",
    "#     for i, x in enumerate(genome[chrom]):\n",
    "#         if x == 'C' or x == 'G':\n",
    "#             cg_arr[i] = True\n",
    "#         elif x == 'A' or x == 'T':\n",
    "#             at_arr[i] = True\n",
    "#     cov_arr = np.load(f'data/coverage/2f1r/big_Col-0-1_{chrom}.npy')\n",
    "#     cg_cov += np.sum(cg_arr * cov_arr)\n",
    "#     at_cov += np.sum(at_arr * cov_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "effect_size = {x:[dict(), dict()] for x in 'cg pericentromere met acr'.split()}\n",
    "# total_effect_size = {x:0 for x in 'cg pericentromere methylated acr'.split()}\n",
    "effect_regions = 'exon intron te promoter intergenic'.split()\n",
    "\n",
    "for r in effect_regions:\n",
    "    # CG content\n",
    "    effect_size['cg'][True][r] = norm_elem_rates['gen']['Col-0'][f'{r} cg']\n",
    "    effect_size['cg'][False][r] = norm_elem_rates['gen']['Col-0'][f'{r} at']\n",
    "    # total_effect_size[r] += norm_elem_rates['gen']['Col-0'][f'{r} cg'] / norm_elem_rates['gen']['Col-0'][f'{r} at']\n",
    "    \n",
    "    # pericentromere\n",
    "    effect_size['pericentromere'][True][r] = norm_elem_rates['gen']['Col-0'][f'{r} pericentromere']\n",
    "    effect_size['pericentromere'][False][r] = norm_elem_rates['gen']['Col-0'][f'{r} distal']\n",
    "    \n",
    "    # methylated\n",
    "    effect_size['met'][True][r] = norm_elem_rates['gen']['Col-0'][f'{r} met']\n",
    "    effect_size['met'][False][r] = norm_elem_rates['gen']['Col-0'][f'{r} unmet']\n",
    "    \n",
    "    # acr\n",
    "    effect_size['acr'][True][r] = norm_elem_rates['gen']['Col-0'][f'{r} acr']\n",
    "    effect_size['acr'][False][r] = norm_elem_rates['gen']['Col-0'][f'{r} nonacr']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(2, 2, sharex=True)\n",
    "for i, effect in enumerate(effect_size):\n",
    "    ax = axs[i // 2][i % 2]\n",
    "    for j, r in enumerate(effect_regions):\n",
    "        ax.plot((0, 1), (math.log2(effect_size[effect][False][r]), math.log2(effect_size[effect][True][r])), marker='o', color=plt.get_cmap('tab10')(j))\n",
    "        ax.set_title(effect + '?', y=0.8)\n",
    "ax.set_xticks([0, 1], labels=['no', 'yes'])\n",
    "axs[0][1].legend(effect_regions, loc='upper left', bbox_to_anchor=(1, 1))\n",
    "axs[0][0].set_ylabel('log2fold relative substitution rate')\n",
    "fig.suptitle('Breakdown of effects by region')\n",
    "fig.savefig('figs/breakdown_of_effects_by_region.svg', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# elem_cov['gen']['Col-0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_effect_size = dict()\n",
    "weights = [elem_counts['gen']['Col-0'][r] for r in effect_regions]\n",
    "for effect in effect_size:\n",
    "    # get the log2fold difference in relative rate between bases with the effect for each region\n",
    "    relative_rates = [math.log2(effect_size[effect][True][r] / effect_size[effect][False][r]) for r in effect_regions]\n",
    "    \n",
    "    # do a weighted average of the log2fold differences (I'm using log2fold because averaging pure rates would inflate the effect of increasing rates)\n",
    "    avg = sum([relative_rates[i] * (weights[i] / sum(weights)) for i in range(len(relative_rates))])\n",
    "    total_effect_size[effect] = 2 ** avg # undo the log2\n",
    "total_effect_size['selection'] = dnds['Ler-0']['ratio']\n",
    "total_effect_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# acr_relative_rate = norm_elem_rates['gen']['Col-0']['ACRs'] / norm_elem_rates['gen']['Col-0']['not ACRs']\n",
    "# met_relative_rate = norm_elem_rates['gen']['Col-0']['non-genic mC'] / norm_elem_rates['gen']['Col-0']['non-genic unmet C']\n",
    "# cg_relative_rate = (cg_muts / cg_cov) / (at_muts / at_cov)\n",
    "# # cg_relative_rate = (((simple_spectra['Col-0']['CA'] + simple_spectra['Col-0']['CG'] + simple_spectra['Col-0']['CT']) / simple_context_freq['C']) /\n",
    "# #                     ((simple_spectra['Col-0']['TA'] + simple_spectra['Col-0']['TC'] + simple_spectra['Col-0']['TG']) / simple_context_freq['T'])) # FIXME need to include indels here\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "simulations = {effect:{chrom:np.ones(chrom_sizes[chrom], dtype=float) for chrom in chrom_sizes} for effect in list(total_effect_size.keys()) + ['no selection', 'all']}\n",
    "\n",
    "for chrom in tqdm(chrom_sizes):\n",
    "    # cg_arr = np.zeros(chrom_sizes[chrom], dtype=bool)\n",
    "    # for i, x in enumerate(genome[chrom]):\n",
    "    #     if x == 'C' or x == 'G':\n",
    "    #         cg_arr[i] = True\n",
    "    simulations['cg'][chrom] = np.ones(chrom_sizes[chrom], dtype=float) * (total_effect_size['cg'] * region_arrays['cg'][chrom] + ~region_arrays['cg'][chrom])\n",
    "    simulations['pericentromere'][chrom] = np.ones(chrom_sizes[chrom], dtype=float) * (total_effect_size['pericentromere'] * region_arrays['pericentromere'][chrom] + ~region_arrays['pericentromere'][chrom])\n",
    "    simulations['met'][chrom] = np.ones(chrom_sizes[chrom], dtype=float) * (total_effect_size['met'] * region_arrays['met c'][chrom] + ~region_arrays['met c'][chrom])\n",
    "    simulations['acr'][chrom] = np.ones(chrom_sizes[chrom], dtype=float) * (total_effect_size['acr'] * region_arrays['acr'][chrom] + ~region_arrays['acr'][chrom])\n",
    "    simulations['selection'][chrom] = np.ones(chrom_sizes[chrom], dtype=float) * (total_effect_size['selection'] * ns_sites[chrom] + ~ns_sites[chrom])\n",
    "    \n",
    "    simulations['no selection'][chrom] = simulations['cg'][chrom] * simulations['pericentromere'][chrom] * simulations['met'][chrom] * simulations['acr'][chrom]\n",
    "    simulations['all'][chrom] = simulations['no selection'][chrom] * simulations['selection'][chrom]\n",
    "    \n",
    "    # simulations[1][chrom] = simulations[0][chrom] * (cg_relative_rate  * cg_arr +                               ~cg_arr)\n",
    "    # simulations[2][chrom] = simulations[1][chrom] * (acr_relative_rate * region_arrays['ACRs'][chrom] +         ~region_arrays['ACRs'][chrom])\n",
    "    # simulations[3][chrom] = simulations[2][chrom] * (met_relative_rate * region_arrays['genic mCG'][chrom] +    ~region_arrays['genic mCG'][chrom]) # FIXME this needs to be all methylated sites\n",
    "    # simulations[3][chrom] = simulations[3][chrom] * (met_relative_rate * region_arrays['non-genic mC'][chrom] + ~region_arrays['non-genic mC'][chrom])\n",
    "    # simulations[4][chrom] = simulations[3][chrom] * (ns_relative_rate  * ns_sites[chrom] +                      ~ns_sites[chrom])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vis_regions = ['exon', 'intron', 'te', 'promoter', 'intergenic'] # FIXME do I need to care about callable coverage here?\n",
    "\n",
    "# get the genome-wide average mutation rate for each simulation\n",
    "sim_uniforms = dict()\n",
    "for effect in simulations:\n",
    "    sim_uniforms[effect] = sum([np.sum(simulations[effect][chrom]) for chrom in chrom_sizes]) / sum(chrom_sizes.values())\n",
    "\n",
    "# get the mutation rate relative to uniform for each region in each simulation\n",
    "sim_rates = {effect:[] for effect in simulations}\n",
    "for effect in simulations:\n",
    "    for region in vis_regions:\n",
    "        rate_sum = sum([np.sum(simulations[effect][chrom] * region_arrays[region][chrom]) for chrom in chrom_sizes])\n",
    "        region_size = sum([np.count_nonzero(region_arrays[region][chrom]) for chrom in chrom_sizes])\n",
    "        sim_rates[effect].append(rate_sum / region_size / sim_uniforms[effect])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # same as above but pretend I sequenced the simulated samples and had to factor in callable coverage\n",
    "# vis_regions = ['exon', 'intron', 'te', 'promoter', 'intergenic'] # FIXME do I need to care about callable coverage here?\n",
    "\n",
    "# # get the genome-wide average mutation rate for each simulation\n",
    "# sim_uniforms = []\n",
    "# for sim in simulations:\n",
    "#     sim_uniforms.append(sum([np.sum(sim[chrom]) for chrom in chrom_sizes]) / sum(chrom_sizes.values()))\n",
    "\n",
    "# # get the mutation rate relative to uniform for each region in each simulation\n",
    "# sim_rates = [[] for x in simulations]\n",
    "# for i, sim in enumerate(simulations):\n",
    "#     for region in vis_regions:\n",
    "#         rate_sum = sum([np.sum(sim[chrom] * region_arrays[region][chrom]) for chrom in chrom_sizes])\n",
    "#         region_size = sum([np.count_nonzero(region_arrays[region][chrom]) for chrom in chrom_sizes])\n",
    "#         sim_rates[i].append(rate_sum / region_size / sim_uniforms[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(x), len(y), len(low_err), len(high_err)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 4))\n",
    "\n",
    "x = [(x + 1) * 5 + 0 for x in range(len(vis_regions))]\n",
    "y = [log2_min(x) for x in sim_rates['no selection']]\n",
    "# print(x)\n",
    "ax.bar(x, y, color=gtools.base_cmap[4])\n",
    "    \n",
    "# for i, t in enumerate(vis_regions):\n",
    "low_err = [log2_min(norm_elem_rates['gen']['Col-0'][r]) - log2_min(max(0.01, norm_elem_rates['low']['Col-0'][r])) for r in vis_regions]\n",
    "high_err = [log2_min(norm_elem_rates['high']['Col-0'][r]) - log2_min(norm_elem_rates['gen']['Col-0'][r]) for r in vis_regions]\n",
    "x = [(x + 1) * 5 + 1 for x in range(len(vis_regions))]\n",
    "y = [log2_min(norm_elem_rates['gen']['Col-0'][r]) for r in vis_regions]\n",
    "ax.bar(x, y, yerr=[low_err, high_err], color=gen_cmap('Col-0'))\n",
    "\n",
    "x = [(x + 1) * 5 + 2 for x in range(len(vis_regions))]\n",
    "y = [log2_min(x) for x in sim_rates['all']]\n",
    "# print(x)\n",
    "ax.bar(x, y, color=gtools.light_cmap[4])\n",
    "\n",
    "low_err = [log2_min(norm_elem_rates['gen']['Ler-0'][r]) - log2_min(max(0.01, norm_elem_rates['low']['Ler-0'][r])) for r in vis_regions]\n",
    "high_err = [log2_min(norm_elem_rates['high']['Ler-0'][r]) - log2_min(norm_elem_rates['gen']['Ler-0'][r]) for r in vis_regions]\n",
    "x = [(x + 1) * 5 + 3 for x in range(len(vis_regions))]\n",
    "y = [log2_min(norm_elem_rates['gen']['Ler-0'][r]) for r in vis_regions]\n",
    "ax.bar(x, y, yerr=[low_err, high_err], color=gen_cmap('Ler-0'))\n",
    "    \n",
    "    \n",
    "ax.set_xticks([(x + 1) * 5 + 3 / 2 for x in range(len(vis_regions))], vis_regions)\n",
    "ax.legend(['simulation no selection', 'untreated', 'simulation with selection', 'Ler-0'])\n",
    "\n",
    "xlim = ax.get_xlim()\n",
    "ax.hlines(0, -10, 100, color='black')\n",
    "ax.set_xlim(xlim)\n",
    "ax.set_ylabel('log2fold relative substitution rate')\n",
    "fig.savefig('figs/simulated_mutation_distributions.svg', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# old version that plotted all simulations\n",
    "# fig, ax = plt.subplots(figsize=(8, 4))\n",
    "# for i, effect in enumerate(simulations):\n",
    "#     color = plt.get_cmap('tab20b')(i + 15) if i > 0 else 'black'\n",
    "#     x = [(x + 1) * (len(simulations) + 2) + i for x in range(len(vis_regions))]\n",
    "#     y = [log2_min(x) for x in sim_rates[effect]]\n",
    "#     # print(x)\n",
    "#     ax.bar(x, y, color=color)\n",
    "    \n",
    "# # for i, t in enumerate(vis_regions):\n",
    "# low_err = [log2_min(norm_elem_rates['gen']['Col-0'][r]) - log2_min(max(0.01, norm_elem_rates['low']['Col-0'][r])) for r in vis_regions]\n",
    "# high_err = [log2_min(norm_elem_rates['high']['Col-0'][r]) - log2_min(norm_elem_rates['gen']['Col-0'][r]) for r in vis_regions]\n",
    "# x = [(x + 1) * (len(simulations) + 2) + len(simulations) for x in range(len(vis_regions))]\n",
    "# y = [log2_min(norm_elem_rates['gen']['Col-0'][r]) for r in vis_regions]\n",
    "# ax.bar(x, y, yerr=[low_err, high_err], color=gen_cmap('Col-0'))\n",
    "\n",
    "# low_err = [log2_min(norm_elem_rates['gen']['Ler-0'][r]) - log2_min(max(0.01, norm_elem_rates['low']['Ler-0'][r])) for r in vis_regions]\n",
    "# high_err = [log2_min(norm_elem_rates['high']['Ler-0'][r]) - log2_min(norm_elem_rates['gen']['Ler-0'][r]) for r in vis_regions]\n",
    "# x = [(x + 1) * (len(simulations) + 2) + (len(simulations) + 1) for x in range(len(vis_regions))]\n",
    "# y = [log2_min(norm_elem_rates['gen']['Ler-0'][r]) for r in vis_regions]\n",
    "# ax.bar(x, y, yerr=[low_err, high_err], color=gen_cmap('Ler-0'))\n",
    "    \n",
    "    \n",
    "# # ax.set_xticks([(x + 1) * (len(simulations) + 2) + (len(simulations) + 2) / 2 for x in range(len(vis_regions))], vis_regions)\n",
    "# ax.legend(list(simulations.keys()) + ['observed Col-0 somatic', 'observed Ler-0 germline'])\n",
    "\n",
    "# xlim = ax.get_xlim()\n",
    "# ax.hlines(0, -10, 100, color='black')\n",
    "# ax.set_xlim(xlim)\n",
    "# ax.set_ylabel('log2fold mutation/polymorphism rate relative to uniform')\n",
    "# fig.savefig('figs/simulated_mutation_distributions.svg', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "low_err"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
